---
title: Supplemental Material for "Multiple colonizations of freshwater by the Gulf pipefish reveal a shared genomic signature of adaptation"
preprint: false
author: 
  - name: Sarah P. Flanagan
    affiliation: 1
    corresponding: true
    email: spflanagan.phd@gmail.com
  - name: Emily Rose
    affiliation: 2
  - name: Adam Jones
    affiliation: 3
affiliation:
  - code: 1
    address: School of Biological Sciences, University of Canterbury, 4800 Private Bag, Christchurch 8140 New Zealand
  - code: 2
    address: Department of Biology, The University of Tampa, Tampa, FL 33606 USA
  - code: 3
    address: Department of Biological Sciences, University of Idaho, Moscow, ID 83844 USA
abstract: >
  This document includes supplementary material for the paper. 
header-includes: >
  \usepackage{lipsum}
bibliography: programs.bib
output:
  pdf_document:
    fig_caption: yes
    keep_tex: yes
    number_sections: yes
    template: manuscript.latex
  html_document: null
  word_document: null
fontsize: 11pt
capsize: normalsize
csl: molecular-ecology.csl
documentclass: article
spacing: doublespacing
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,out.extra='',fig.pos="H")
knitr::opts_knit$set(root.dir='../fwsw_results/')
```
```{r source}
source("../../gwscaR/R/gwscaR.R")
source("../../gwscaR/R/gwscaR_plot.R")
source("../../gwscaR/R/gwscaR_utility.R")
source("../../gwscaR/R/gwscaR_fsts.R")
source("../../gwscaR/R/gwscaR_popgen.R")
source("../../gwscaR/R/vcf2dadi.R")
source("../R/250_dadi_analysis.R")
library(knitr)
pop.list<-c("ALFW","ALST","FLCC","FLLG","LAFW","TXCC","TXFW")
```

Some useful functions are in the `250_dadi_analysis.R` script.

# Create a dadi SNPs file from the vcf {-}

```{r create_dadi, eval=FALSE}
vcf<-parse.vcf("stacks/populations_subset75/batch_2.vcf")
dadi<-vcf2dadiSNPs(vcf,pop.list = pop.list,filename = "dadi_analysis/fwsw.dadi.snps")
dadi<-read.delim("dadi_analysis/fwsw.dadi.snps")
projections<-unlist(lapply(pop.list,function(pop){ n<-length(grep(pop,colnames(vcf)))}))*2


names(projections)<-pop.list
best.pops<-c("TXFW","TXCB","LAFW","ALST","ALFW","FLLG","FLCC","TXFW ","TXCB ","LAFW ","ALST ","ALFW ","FLLG ","FLCC ")
best.pops<-c("TXFW","TXCB","LAFW","ALST","ALFW","FLLG","FLCC","TXFW.1","TXCB.1","LAFW.1","ALST.1","ALFW.1","FLLG.1","FLCC.1")
dadi<-dadi[,c("Ingroup","Outgroup","Allele1",best.pops[1:7],"Allele2",best.pops[8:14],"GeneID","Position")]
dim(dadi[rowSums(dadi[,colnames(dadi) %in% best.pops])==sum(projections[best.pops[1:7]]),])

#transform this and remove everything but population info for ease of operations
dd<-as.data.frame(cbind(colnames(dadi)[colnames(dadi) %in% best.pops],t(dadi[,colnames(dadi) %in% best.pops])),row.names = 0,stringsAsFactors = FALSE)
#rename the pops to be factors, essentially
dd$V1<-gsub("\\.1","",dd$V1)
#calculate sums per locus for each population
ns<-data.frame(do.call(rbind,lapply(2:ncol(dd),function(c){ tapply(dd[,c],dd$V1,function(x){ sum(as.numeric(x)) }) })))
rownames(ns)<-dadi$GeneID
nozeros<-ns[!(apply(ns, 1, function(y) any(y == 0))),] #512 had zeroes
props<-apply(nozeros,2,function(x) { p<-x/max(x) } )
keepers<-props[apply(props, 1, function(y) any(y >=0.75)),]

#now save these in the dadi format
write.table(dadi[dadi$GeneID %in% rownames(keepers),],"dadi_analysis/fwsw.dadi.pruned.snps",quote=FALSE,row.names = FALSE,
            col.names = c("Ingroup","Outgroup","Allele1","TXFW","TXCB","LAFW","ALST","ALFW","FLLG","FLCC","Allele2","TXFW","TXCB","LAFW","ALST","ALFW","FLLG","FLCC","GeneID","Position"))
```

```{r check_dadis, eval=FALSE}
vcf.files<-c("stacks/populations_subset50/batch_2.vcf",
             "stacks/populations_subset75/batch_2.vcf",
             #"stacks/populations_subset85/batch_2.vcf",
             #"stacks/populations_subset90/batch_2.vcf",
             #"stacks/populations_subset95/batch_2.vcf",
             "stacks/populations_subset100/batch_2.vcf")

dadis<-lapply(vcf.files,function(file){
  full.vcf<-parse.vcf(file)
  vcf<-choose.one.snp(full.vcf)
  vcf.out.name<-gsub("batch_2.vcf","batch_2.pruned.vcf",file)
  write.table(x = vcf,file = vcf.out.name,quote = FALSE,sep='\t',row.names = FALSE,col.names = TRUE)
  dadi.out<-paste("dadi_analysis/fwsw",gsub(".*populations_subset(\\d+).*","\\1",file),".dadi.snps",sep="")
  pop.list<-c("ALFW","ALST","FLCC","FLLG","LAFW","TXCC","TXFW")
  dadi<-vcf2dadiSNPs(vcf,pop.list = pop.list,filename = dadi.out)
  return(dadi)
})

#compare
kable(cbind(file=vcf.files,numer_snps=unlist(lapply(dadis,nrow))))
```

TODO: Write code to read in the files and compare

I'm going to use the 75% of individuals dataset.

## Running $\delta_A\delta_I$ {-}

I ran the optimize functions from (dportik's $\delta_A\delta_I$ pipeline)[https://github.com/dportik/dadi_pipeline] for each pairwise combination of the 7 populations with the 26 models from Rougeux. These ran using python2.7 (because I used $\delta_A\delta_I$ v. xxx) using individual scripts for each population pair and each model (in dadi_scripts/). These scripts were generated using `252_generate_dadi_scripts.sh`. Each of the population-pair-model scripts runs `252_pairwise_dadi.py`, which requires the dadi-pipeline scripts and the file `rougeux_models.py`, which contains the 26 models run here. Each script was run using either `252_run_dadi_scripts.sh` or `run_dadi_scripts_abacus.sh`, depending on whether the scripts were run on a local machine (C001KR or rccuser\@10.195.0.46) or on the cluster (abacus). 

Single population plots were generated using `250_dadiPlots.py`.

To run the models I saved all of the population-pair-model combinations to a file to make it easier to run specific combinations.


```{r modelCombos,eval=FALSE}
pops=c('FLLG', 'FLCC', 'ALFW', 'ALST', 'LAFW', 'TXFW', 'TXCC')   #
models=c('SI', 'IM', 'AM', 'SC', 'SI2N', 'SIG', 'SI2NG', 'IMG', 'IM2N', 'IM2m', 'IM2NG', 'IM2mG', 'AM2N', 'AMG', 'AM2m', 'AM2NG', 'AM2N2m', 'AM2mG', 'AM2N2mG', 'SCG' ,'SC2N', 'SC2m', 'SC2NG', 'SC2N2m' ,'SC2mG' ,'SC2N2mG')
 



############ CREATE POP COMBOS TO RUN ############
tasks=NULL
for(i in 1:(length(pops)-1)){
  for(j in (i+1):length(pops)){
    for(mod in 1:length(models)){
      tasks<-c(tasks,paste(pops[i],pops[j],models[mod],sep="_"))
    }
  }
}

write.csv(tasks,"model_list_dadi.csv",row.names = TRUE,quote=FALSE)
```


## Analyzing $\delta_A\delta_I$ output {-}

### FLLG vs FLCC


```{r}
flcc_fllg_opts<-dadi.modelcomp(path = "dadi_results/FLLG_FLCC/",
                               pattern="*optimized.*",id="FLLG_FLCC")
flcc_fllg<-modelComparison(flcc_fllg_opts)
plot(as.numeric(flcc_fllg_opts$log.likelihood)~
              as.factor(flcc_fllg_opts$Model),las=3,
        pch=19,xlab="Model",ylab="Maximum log likelihood")

```
It's looking like IM2mG model has a higher on average log likelihood than SC2mG. Let's see if the IM2mG parameter distributions are better

```{r}
flcc_fllg_sc2mg<-do.call(rbind,extract_dadi_params(flcc_fllg_opts[flcc_fllg_opts$Model=="SC2mG",]))
flcc_fllg_im2mg<-do.call(rbind,extract_dadi_params(flcc_fllg_opts[flcc_fllg_opts$Model=="IM2mG",]))
```

```{r}
par(mfrow=c(1,2))
boxplot(flcc_fllg_sc2mg[,11:22])
points(1:12,
       flcc_fllg_sc2mg[which.max(flcc_fllg_sc2mg$log.likelihood),11:22],
       pch=8,col="red",lwd=2)

boxplot(flcc_fllg_im2mg[,11:21])
points(1:11,
       flcc_fllg_im2mg[which.max(flcc_fllg_im2mg$log.likelihood),11:21],
       pch=8,col="red",lwd=2)
```

#### SC2mG

```{r}
flcc_fllg_sc2mg<-do.call(rbind,extract_dadi_params(flcc_fllg_opts[flcc_fllg_opts$Model=="SC2mG",]))
```

```{r}
boxplot(flcc_fllg_sc2mg[,11:22])
points(1:12,
       flcc_fllg_sc2mg[which.max(flcc_fllg_sc2mg$log.likelihood),11:22],
       pch=8,col="red",lwd=2)
```




