---
title: Supplemental Material for "Multiple colonizations of freshwater by the Gulf pipefish reveal a shared genomic signature of adaptation"
preprint: false
author: 
  - name: Sarah P. Flanagan
    affilnum: 1
    corresponding: true
    email: spflanagan.phd@gmail.com
  - name: Emily Rose
    affilnum: 2
  - name: Adam Jones
    affilnum: 3
affiliation:
  - affilnum: 1
    affil: School of Biological Sciences, University of Canterbury, 4800 Private Bag, Christchurch 8140 New Zealand
  - affilnum: 2
    affil: Department of Biology, The University of Tampa, Tampa, FL 33606 USA
  - affilnum: 3
    affil: Department of Biological Sciences, University of Idaho, Moscow, ID 83844 USA
abstract: >
  This document includes supplementary material for the paper. 
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
bibliography: programs.bib
output:
  pdf_document:
    fig_caption: yes
    keep_tex: yes
    number_sections: no
    toc: true
    toc_depth: 2
    template: manuscript.latex
  html_document: null
  word_document: null
fontsize: 11pt
capsize: normalsize
csl: molecular-ecology.csl
documentclass: article
spacing: singlespacing
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,out.extra='',fig.pos="H",
                      warning=FALSE,message=FALSE)
knitr::opts_knit$set(root.dir='../fwsw_results/')
```
```{r source}
source("../../gwscaR/R/gwscaR.R")
source("../../gwscaR/R/gwscaR_plot.R")
source("../../gwscaR/R/gwscaR_utility.R")
source("../../gwscaR/R/gwscaR_fsts.R")
source("../../gwscaR/R/gwscaR_popgen.R")
source("../../gwscaR/R/vcf2dadi.R")
source("../R/250_dadi_analysis.R")
library(knitr); library(kableExtra)
library(png)
library(scales)
pop.list<-c("ALFW","ALST","FLCC","FLLG","LAFW","TXCC","TXFW")
```

Some useful functions are in the `250_dadi_analysis.R` script.

# Create a dadi SNPs file from the vcf {-}

```{r create_dadi, eval=FALSE}
vcf<-parse.vcf("stacks/populations_subset75/batch_2.vcf")
dadi<-vcf2dadiSNPs(vcf,pop.list = pop.list,filename = "dadi_analysis/fwsw.dadi.snps")
dadi<-read.delim("dadi_analysis/fwsw.dadi.snps")
projections<-unlist(lapply(pop.list,function(pop){ 
  n<-length(grep(pop,colnames(vcf)))
}))*2


names(projections)<-pop.list
best.pops<-c("TXFW","TXCB","LAFW","ALST","ALFW","FLLG","FLCC",
             "TXFW ","TXCB ","LAFW ","ALST ","ALFW ","FLLG ","FLCC ")
best.pops<-c("TXFW","TXCB","LAFW","ALST","ALFW","FLLG","FLCC",
             "TXFW.1","TXCB.1","LAFW.1","ALST.1","ALFW.1","FLLG.1","FLCC.1")
dadi<-dadi[,c("Ingroup","Outgroup","Allele1",best.pops[1:7],
              "Allele2",best.pops[8:14],"GeneID","Position")]
dim(dadi[rowSums(dadi[,colnames(dadi) %in% best.pops])==
           sum(projections[best.pops[1:7]]),])

#transform this and remove everything but population info for ease of operations
dd<-as.data.frame(cbind(colnames(dadi)[colnames(dadi) %in% best.pops],
                        t(dadi[,colnames(dadi) %in% best.pops])),
                  row.names = 0,stringsAsFactors = FALSE)
#rename the pops to be factors, essentially
dd$V1<-gsub("\\.1","",dd$V1)
#calculate sums per locus for each population
ns<-data.frame(do.call(rbind,lapply(2:ncol(dd),function(c){ 
  tapply(dd[,c],dd$V1,function(x){ sum(as.numeric(x)) }) })))
rownames(ns)<-dadi$GeneID
nozeros<-ns[!(apply(ns, 1, function(y) any(y == 0))),] #512 had zeroes
props<-apply(nozeros,2,function(x) { p<-x/max(x) } )
keepers<-props[apply(props, 1, function(y) any(y >=0.75)),]

#now save these in the dadi format
write.table(dadi[dadi$GeneID %in% rownames(keepers),],
            "dadi_analysis/fwsw.dadi.pruned.snps",
            quote=FALSE,row.names = FALSE,
            col.names = c("Ingroup","Outgroup","Allele1",
                          "TXFW","TXCB","LAFW","ALST","ALFW","FLLG","FLCC",
                          "Allele2","TXFW","TXCB","LAFW","ALST","ALFW",
                          "FLLG","FLCC","GeneID","Position"))
```

```{r check_dadis, eval=FALSE}
vcf.files<-c("stacks/populations_subset50/batch_2.vcf",
             "stacks/populations_subset75/batch_2.vcf",
             #"stacks/populations_subset85/batch_2.vcf",
             #"stacks/populations_subset90/batch_2.vcf",
             #"stacks/populations_subset95/batch_2.vcf",
             "stacks/populations_subset100/batch_2.vcf")

dadis<-lapply(vcf.files,function(file){
  full.vcf<-parse.vcf(file)
  vcf<-choose.one.snp(full.vcf)
  vcf.out.name<-gsub("batch_2.vcf","batch_2.pruned.vcf",file)
  write.table(x = vcf,file = vcf.out.name,quote = FALSE,
              sep='\t',row.names = FALSE,col.names = TRUE)
  dadi.out<-paste("dadi_analysis/fwsw",
                  gsub(".*populations_subset(\\d+).*","\\1",file),
                  ".dadi.snps",sep="")
  pop.list<-c("ALFW","ALST","FLCC","FLLG","LAFW","TXCC","TXFW")
  dadi<-vcf2dadiSNPs(vcf,pop.list = pop.list,filename = dadi.out)
  return(dadi)
})

#compare
kable(cbind(file=vcf.files,numer_snps=unlist(lapply(dadis,nrow))))
```

TODO: Write code to read in the files and compare

I'm going to use the 75% of individuals dataset.

## Running dadi {-}

I ran the optimize functions from (dportik's $\delta_A\delta_I$ pipeline)[https://github.com/dportik/dadi_pipeline] for each pairwise combination of the 7 populations with the 26 models from Rougeux. These ran using python2.7 (because I used $\delta_A\delta_I$ v. xxx) using individual scripts for each population pair and each model (in dadi_scripts/). These scripts were generated using `252_generate_dadi_scripts.sh`. Each of the population-pair-model scripts runs `252_pairwise_dadi.py`, which requires the dadi-pipeline scripts and the file `rougeux_models.py`, which contains the 26 models run here. Each script was run using either `252_run_dadi_scripts.sh` or `run_dadi_scripts_abacus.sh`, depending on whether the scripts were run on a local machine (C001KR or rccuser\@10.195.0.46) or on the cluster (abacus).

This is a visualization of all of the different models.

```{r dadiModels,fig.height=8,fig.width=6,dpi=300,dev='png',fig.path="../figs/", fig.cap="Schematic diagrams of each model, color-coded by base model (Strict Isolation, Isolation with Migration, Ancient Migration, or Secondary Contact). 'Linked Sel.' refers to linked selection, where a sub-section of the genome experienced a different effective population size than the remainder of the genome (hrf). 'Het. Migration' refers to differential introgression, where some regions of the genome experienced different migration rates than others. Each row demonstrates how parameters were added to the base models. Strict Isolation models contained now migration, so only linked selection and growth could be incorporated. Under Isolation with Migration scenarios, differentiating between linked selection and heterogeneous migration is difficult, so the models that combined both were not included for the Isolation with Migration scenarios."}
figPath<-"../figs/dadi_models/"
mods<-c("SI","IM","AM","SC")
adds<-c("","G","2N","2NG","2m","2mG","2N2m","2N2mG")

par(mfrow=c(8,4),oma=c(1,6,4,1),mar=c(0,0,0,0))

for(j in 1:length(adds)){
  for(i in 1:length(mods)){
    plot(c(0,1),c(0,1),type='n',axes=FALSE)
    imgName<-paste0(figPath,mods[i],adds[j],".PNG")
    if(file.exists(imgName)){
      img<-readPNG(imgName)
      rasterImage(img,0,0,1,1,xpd=TRUE)
    }
  }
}

# add headers
mtext("Strict
Isolation",side=3,outer=TRUE,at=0.125,font=2)
mtext("Isolation w/ 
Migration",side=3,outer=TRUE,at=0.375,font=2)
mtext("Ancient
Migration",side=3,outer=TRUE,at=0.62,font=2)
mtext("Secondary
Contact",side=3,outer=TRUE,at=0.875,font=2)

# add rownames
mtext("Growth",side=2,outer=TRUE,at=(7/8)-0.05,font=2,las=1,adj=0.65)
mtext("Linked Sel.",side=2,outer=TRUE,at=(6/8)-0.05,font=2,las=1,adj=0.65)
mtext("Linked Sel.+
Growth",side=2,outer=TRUE,at=(5/8)-0.05,font=2,las=1,adj=0.65)
mtext("Het. Migration",side=2,outer=TRUE,at=(4/8)-0.05,font=2,las=1,adj=0.65)
mtext("Het. Migration+
Growth",side=2,outer=TRUE,at=(3/8)-0.05,font=2,las=1,adj=0.65)
mtext("Linked Sel.+
Het. Migration",side=2,outer=TRUE,at=(2/8)-0.05,font=2,las=1,adj=0.65)
mtext("Linked Sel.+
Het. Migration+
Growth",side=2,outer=TRUE,at=(1/8)-0.05,font=2,las=1,adj=0.65)

```


Single population plots were generated using `250_dadiPlots.py`.

To run the models I saved all of the population-pair-model combinations to a file to make it easier to run specific combinations.


```{r modelCombos,eval=FALSE}
pops=c('FLLG', 'FLCC', 'ALFW', 'ALST', 'LAFW', 'TXFW', 'TXCC')   #
models=c('SI', 'IM', 'AM', 'SC', 'SI2N', 'SIG', 'SI2NG', 'IMG', 
         'IM2N', 'IM2m', 'IM2NG', 'IM2mG', 'AM2N', 'AMG', 'AM2m', 
         'AM2NG', 'AM2N2m', 'AM2mG', 'AM2N2mG', 'SCG' ,'SC2N', 
         'SC2m', 'SC2NG', 'SC2N2m' ,'SC2mG' ,'SC2N2mG')
 



############ CREATE POP COMBOS TO RUN ############
tasks=NULL
for(i in 1:(length(pops)-1)){
  for(j in (i+1):length(pops)){
    for(mod in 1:length(models)){
      tasks<-c(tasks,paste(pops[i],pops[j],models[mod],sep="_"))
    }
  }
}

write.csv(tasks,"model_list_dadi.csv",row.names = TRUE,quote=FALSE)
```



## Analyzing dadi output {-}

```{r dadi_InitialCompare}
dadi_initialCompare<-function(comparison,modSel="max"){
  opts<-dadi.modelcomp(path = paste0("dadi_results/",comparison,"/"),
                               pattern="*optimized.*",id=comparison)
  comp<-modelComparison(opts,modSel)
  Ns<-tapply(opts$log.likelihood,opts$Model,length)
  plot(as.numeric(opts$log.likelihood)~
              as.factor(opts$Model),las=3,
       ylim=c(min(as.numeric(opts$log.likelihood)-50),
              max(as.numeric(opts$log.likelihood)+50)),
        pch=19,xlab="",ylab="Maximum log likelihood")
  text(x=1:length(Ns),
       y=rep(min(as.numeric(opts$log.likelihood)-50),length(Ns)),
       Ns,cex=0.75)
  return(list(opts,comp))
}
```


### FLFW vs FLCC


```{r fllgOpts, fig.cap="Comparison of log likelihoods for FLFW-FLCC models"}
fllg_flcc<-dadi_initialCompare("FLLG_FLCC",modSel="median")
```
```{r fffwTable}
sumTab<-fllg_flcc[[2]]
sumTab<-sumTab[,-2]
sumTab<-sumTab[,c(9,1:5,10:12,6,7)]
sumTab$parms<-apply(sumTab,1,function(dat){
  return(paste0(unlist(strsplit(dat["params"],",")),"=",
         unlist(strsplit(dat["optimized_params"],","))))
})
sumTab<-sumTab[,-c(10,11)]  

landscape(kable(sumTab,booktabs=TRUE,
      caption = "Table showing the results of the run with the median log likelihood for each model",
      row.names = FALSE)) %>%
  kable_styling(latex_options="scale_down")
```



It's looking like IM2mG model has a higher on average log likelihood than SC2mG. Let's see if the IM2mG parameter distributions are better

```{r fllgBestMods}
flcc_fllg_sc2mg<-do.call(rbind,
                         extract_dadi_params(fllg_flcc[[1]][fllg_flcc[[1]]$Model==
                                                              "SC2mG",]))
flcc_fllg_im2mg<-do.call(rbind,
                         extract_dadi_params(fllg_flcc[[1]][fllg_flcc[[1]]$Model==
                                                              "IM2mG",]))
```

```{r fllgCompareMods, fig.cap="Boxplot showing the parameter estimates from the replicates of the SC2mG model (left) and the IM2mG model (right) for the FLFW vs FLCC pairwise comparison. The red stars represent the model parameters from the run with the highest log likelihood."}
par(mfrow=c(1,2))
boxplot(flcc_fllg_sc2mg[,11:22],main="SC2mG")
points(1:12,
       flcc_fllg_sc2mg[which.max(flcc_fllg_sc2mg$log.likelihood),11:22],
       pch=8,col="red",lwd=2)

boxplot(flcc_fllg_im2mg[,11:21],main="IM2mG")
points(1:11,
       flcc_fllg_im2mg[which.max(flcc_fllg_im2mg$log.likelihood),11:21],
       pch=8,col="red",lwd=2)
```
```{r setupOutTable}
outNames<-colnames(do.call(rbind,
                           extract_dadi_params(
                             fllg_flcc[[1]][fllg_flcc[[1]]$Model=="SC2N2mG",])))
outNames<-outNames[!outNames %in% c("Replicate","params","optimized_params","rank","id")]
outNames<-c("PopPair",outNames,"Tam")

outTable<-data.frame(matrix(nrow = 1,ncol=length(outNames),
                            dimnames = list("",outNames)))
```


```{r saveBest1}
best<-flcc_fllg_im2mg[which.max(flcc_fllg_im2mg$log.likelihood),]

outTable[1,1]<-c("FLFW_FLCC")
outTable[1,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]

bestModels<-data.frame(Pop1="FLFW",
                       Pop2="FLCC",
                       Model="IM2mG",
                       stringsAsFactors = FALSE)
```

### FLFW vs ALFW

```{r ffafOpts, fig.cap="Comparison of log likelihoods for FLFW-ALFW models" }
fllg_alfw<-dadi_initialCompare("FLLG_ALFW")
```
```{r BestMods2}
fllg_alfw_am2ng<-do.call(rbind,
                         extract_dadi_params(fllg_alfw[[1]][fllg_alfw[[1]]$Model==
                                                              "AM2NG",]))
```

```{r saveBest2}
best<-fllg_alfw_am2ng[which.max(fllg_alfw_am2ng$log.likelihood),]

outTable[2,1]<-c("FLFW_ALFW")
outTable[2,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]


bestModels[2,]<-c("FLFW","ALFW",Model="AM2NG")
```

### FLFW vs ALST

```{r ffasOpts, fig.cap="Comparison of log likelihoods for FLFW-ALST models"}
fllg_alst<-dadi_initialCompare("FLLG_ALST")
```
```{r BestMods3}
fllg_alst_am2mg<-do.call(rbind,
                         extract_dadi_params(fllg_alst[[1]][fllg_alst[[1]]$Model==
                                                              "AM2mG",]))
```

```{r saveBest3}
best<-fllg_alst_am2mg[which.max(fllg_alst_am2mg$log.likelihood),]

outTable[3,1]<-c("FLFW_ALST")
outTable[3,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]


bestModels[3,]<-c("FLFW","ALST",Model="AM2mG")
```

### FLFW vs LAFW

```{r fflfOpts, fig.cap="Comparison of log likelihoods for FLFW-LAFW models"}
fllg_lafw<-dadi_initialCompare("FLLG_LAFW")
```
```{r BestMods4}
fllg_lafw_SC2N2mG<-do.call(rbind,
                         extract_dadi_params(fllg_lafw[[1]][fllg_lafw[[1]]$Model==
                                                              "SC2N2mG",]))
```
```{r saveBest4}
best<-fllg_lafw_SC2N2mG[which.max(fllg_lafw_SC2N2mG$log.likelihood),]

outTable[4,1]<-c("FLFW_LAFW")
outTable[4,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]


bestModels[4,]<-c("FLFW","LAFW",Model="SC2N2mG")
```

### FLFW vs TXFW

```{r fftfOpts, fig.cap="Comparison of log likelihoods for FLFW-TXFW models"}
fllg_txfw<-dadi_initialCompare("FLLG_TXFW")
```
```{r BestMods5}
fllg_txfw_AM2N2mG<-do.call(rbind,
                         extract_dadi_params(fllg_txfw[[1]][fllg_txfw[[1]]$Model==
                                                              "AM2N2mG",]))
```
```{r saveBest5}
best<-fllg_txfw_AM2N2mG[which.max(fllg_txfw_AM2N2mG$log.likelihood),]

outTable[5,1]<-c("FLFW_TXFW")
outTable[5,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]

bestModels[5,]<-c("FLFW","TXFW",Model="AM2N2mG")
```

### FLFW vs TXCC

```{r fftsOpts, fig.cap="Comparison of log likelihoods for FLFW-TXCC models"}
fllg_txcc<-dadi_initialCompare("FLLG_TXCC")
```
```{r BestMods6}
fllg_txcc_AM2mG<-do.call(rbind,
                         extract_dadi_params(fllg_txcc[[1]][fllg_txcc[[1]]$Model==
                                                              "AM2mG",]))
```
```{r saveBest6}
best<-fllg_txcc_AM2mG[which.max(fllg_txcc_AM2mG$log.likelihood),]

outTable[6,1]<-c("FLFW_TXCC")
outTable[6,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]

bestModels[6,]<-c("FLFW","TXCC",Model="AM2mG")
```

### FLCC vs ALFW

```{r fsafOpts, fig.cap="Comparison of log likelihoods for FLCC-ALFW models"}
flcc_alfw<-dadi_initialCompare("FLCC_ALFW")
```
```{r BestMods7}
flcc_alfw_IM<-do.call(rbind,
                         extract_dadi_params(flcc_alfw[[1]][flcc_alfw[[1]]$Model==
                                                              "IM",]))
```
```{r saveBest7}
best<-flcc_alfw_IM[which.max(flcc_alfw_IM$log.likelihood),]

outTable[7,1]<-c("FLCC_ALFW")
outTable[7,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]

bestModels[7,]<-c("FLCC","ALFW",Model="IM")
```

### FLCC vs ALST

```{r fsasOpts, fig.cap="Comparison of log likelihoods for FLCC-ALST models"}
flcc_alst<-dadi_initialCompare("FLCC_ALST")
```
```{r BestMods8}
flcc_alst_sc2n<-do.call(rbind,
                      extract_dadi_params(
                        flcc_alst[[1]][flcc_alst[[1]]$Model=="SC2N",]))
```
```{r saveBest8}
best<-flcc_alst_sc2n[which.max(flcc_alst_sc2n$log.likelihood),]

outTable[8,1]<-c("FLCC_ALST")
outTable[8,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]


bestModels[8,]<-c("FLCC","ALST",Model="SC2N")
```

### FLCC vs LAFW

```{r fslfOpts, fig.cap="Comparison of log likelihoods for FLCC-LAFW models"}
flcc_lafw<-dadi_initialCompare("FLCC_LAFW")
```
```{r BestMods9}
flcc_lafw_im2mg<-do.call(rbind,
                      extract_dadi_params(
                        flcc_lafw[[1]][flcc_lafw[[1]]$Model=="IM2mG",]))
```
```{r saveBest9}
best<-flcc_lafw_im2mg[which.max(flcc_lafw_im2mg$log.likelihood),]

outTable[9,1]<-c("FLCC_LAFW")
outTable[9,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]

bestModels[9,]<-c("FLCC","LAFW",Model="IM2mG")
```

### FLCC vs TXFW

```{r fstfOpts, fig.cap="Comparison of log likelihoods for FLCC-TXFW models"}
flcc_txfw<-dadi_initialCompare("FLCC_TXFW")
```
```{r BestMods10}
flcc_txfw_im2mg<-do.call(rbind,
                      extract_dadi_params(
                        flcc_txfw[[1]][flcc_txfw[[1]]$Model=="IM2mG",]))
```
```{r saveBest10}
best<-flcc_txfw_im2mg[which.max(flcc_txfw_im2mg$log.likelihood),]

outTable[10,1]<-c("FLCC_TXFW")
outTable[10,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]

bestModels[10,]<-c("FLCC","TXFW",Model="IM2mG")
```

### FLCC vs TXCC

```{r fstsOpts, fig.cap="Comparison of log likelihoods for FLCC-TXCC models"}
flcc_txcc<-dadi_initialCompare("FLCC_TXCC")
```
```{r BestMods11}
flcc_txcc_im2mg<-do.call(rbind,
                      extract_dadi_params(
                        flcc_txcc[[1]][flcc_txcc[[1]]$Model=="IM2mG",]))
```
```{r saveBest11}
best<-flcc_txcc_im2mg[which.max(flcc_txcc_im2mg$log.likelihood),]

outTable[11,1]<-c("FLCC_TXCC")
outTable[11,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]

bestModels[11,]<-c("FLCC","TXCC",Model="IM2mG")
```

### ALFW vs ALST

```{r afasOpts, fig.cap="Comparison of log likelihoods for ALFW-ALST models"}
alfw_alst<-dadi_initialCompare("ALFW_ALST")
```
```{r BestMods12}
alfw_alst_sc2m<-do.call(rbind,
                      extract_dadi_params(
                        alfw_alst[[1]][alfw_alst[[1]]$Model=="SC2m",]))
```
```{r saveBest12}
best<-alfw_alst_sc2m[which.max(alfw_alst_sc2m$log.likelihood),]

outTable[12,1]<-c("ALFW_ALST")
outTable[12,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]

bestModels[12,]<-c("ALFW","ALST",Model="SC2m")
```

### ALFW vs LAFW

```{r aflfOpts, fig.cap="Comparison of log likelihoods for ALFW-LAFW models"}
alfw_lafw<-dadi_initialCompare("ALFW_LAFW")
```
```{r BestMods13}
alfw_lafw_sc2n2mg<-do.call(rbind,
                      extract_dadi_params(
                        alfw_lafw[[1]][alfw_lafw[[1]]$Model=="SC2N2mG",]))
```
```{r saveBest13}
best<-alfw_lafw_sc2n2mg[which.max(alfw_lafw_sc2n2mg$log.likelihood),]

outTable[13,1]<-c("ALFW_LAFW")
outTable[13,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]

bestModels[13,]<-c("ALFW","LAFW",Model="SC2N2mG")
```

### ALFW vs TXFW

```{r aftfOpts, fig.cap="Comparison of log likelihoods for ALFW-TXFW models"}
alfw_txfw<-dadi_initialCompare("ALFW_TXFW")
```
```{r BestMods14}
alfw_txfw_im2n<-do.call(rbind,
                      extract_dadi_params(
                        alfw_txfw[[1]][alfw_txfw[[1]]$Model=="IM2N",]))
```
```{r saveBest14}
best<-alfw_txfw_im2n[which.max(alfw_txfw_im2n$log.likelihood),]

outTable[14,1]<-c("ALFW_TXFW")
outTable[14,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]

bestModels[14,]<-c("ALFW","TXFW",Model="IM2N")
```


### ALFW vs TXCC

```{r aftsOpts, fig.cap="Comparison of log likelihoods for ALFW-TXCC models"}
alfw_txcc<-dadi_initialCompare("ALFW_TXCC")
```
```{r BestMods15}
alfw_txcc_sc2n2mg<-do.call(rbind,
                      extract_dadi_params(
                        alfw_txcc[[1]][alfw_txcc[[1]]$Model=="SC2N2mG",]))
```
```{r saveBest15}
best<-alfw_txcc_sc2n2mg[which.max(alfw_txcc_sc2n2mg$log.likelihood),]

outTable[15,1]<-c("ALFW_TXCC")
outTable[15,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]

bestModels[15,]<-c("ALFW","TXCC",Model="SC2N2mG")
```


### ALST vs LAFW

```{r aslfOpts, fig.cap="Comparison of log likelihoods for ALST-LAFW models"}
alst_lafw<-dadi_initialCompare("ALST_LAFW")
```
```{r BestMods16}
alst_lafw_im2ng<-do.call(rbind,
                      extract_dadi_params(
                        alst_lafw[[1]][alst_lafw[[1]]$Model=="IM2NG",]))
```
```{r saveBest16}
best<-alst_lafw_im2ng[which.max(alst_lafw_im2ng$log.likelihood),]

outTable[16,1]<-c("ALST_LAFW")
outTable[16,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]

bestModels[16,]<-c("ALST","LAFW",Model="IM2NG")
```


### ALST vs TXFW

```{r astfOpts, fig.cap="Comparison of log likelihoods for ALST-TXFW models"}
alst_txfw<-dadi_initialCompare("ALST_TXFW")
```
```{r BestMods17}
alst_txfw_im<-do.call(rbind,
                      extract_dadi_params(
                        alst_txfw[[1]][alst_txfw[[1]]$Model=="IM",]))
```
```{r saveBest17}
best<-alst_txfw_im[which.max(alst_txfw_im$log.likelihood),]

outTable[17,1]<-c("ALST_TXFW")
outTable[17,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]

bestModels[17,]<-c("ALST","TXFW",Model="IM")
```


### ALST vs TXCC

```{r astsOpts, fig.cap="Comparison of log likelihoods for ALST-TXCC models"}
alst_txcc<-dadi_initialCompare("ALST_TXCC")
```
```{r BestMods18}
alst_txcc_am2mg<-do.call(rbind,
                      extract_dadi_params(
                        alst_txcc[[1]][alst_txcc[[1]]$Model=="AM2mG",]))
```
```{r saveBest18}
best<-alst_txcc_am2mg[which.max(alst_txcc_am2mg$log.likelihood),]

outTable[18,1]<-c("ALST_TXCC")
outTable[18,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]

bestModels[18,]<-c("ALST","TXCC",Model="AM2mG")
```


### LAFW vs TXFW

```{r lftfOpts, fig.cap="Comparison of log likelihoods for LAFW-TXFW models"}
lafw_txfw<-dadi_initialCompare("LAFW_TXFW")
```
```{r BestMods19}
lafw_txfw_scg<-do.call(rbind,
                      extract_dadi_params(
                        lafw_txfw[[1]][lafw_txfw[[1]]$Model=="SCG",]))
```
```{r saveBest19}
best<-lafw_txfw_scg[which.max(lafw_txfw_scg$log.likelihood),]

outTable[19,1]<-c("LAFW_TXFW")
outTable[19,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]

bestModels[19,]<-c("LAFW","TXFW",Model="SCG")
```


### LAFW vs TXCC

```{r lftsOpts, fig.cap="Comparison of log likelihoods for LAFW-TXCC models"}
lafw_txcc<-dadi_initialCompare("LAFW_TXCC")
```
```{r BestMods20}
lafw_txcc_sc2ng<-do.call(rbind,
                      extract_dadi_params(
                        lafw_txcc[[1]][lafw_txcc[[1]]$Model=="SC2NG",]))
```
```{r saveBest20}
best<-lafw_txcc_sc2ng[which.max(lafw_txcc_sc2ng$log.likelihood),]

outTable[20,1]<-c("LAFW_TXCC")
outTable[20,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]

bestModels[20,]<-c("LAFW","TXCC",Model="SC2NG")
```


### TXFW vs TXCC

```{r tftsOpts, fig.cap="Comparison of log likelihoods for TXFW-TXCC models"}
txfw_txcc<-dadi_initialCompare("TXFW_TXCC")
```
```{r BestMods21}
txfw_txcc_am2n2mg<-do.call(rbind,
                      extract_dadi_params(
                        txfw_txcc[[1]][txfw_txcc[[1]]$Model=="AM2N2mG",]))
```
```{r saveBest21}
best<-txfw_txcc_am2n2mg[which.max(txfw_txcc_am2n2mg$log.likelihood),]

outTable[21,1]<-c("TXFW_TXCC")
outTable[21,colnames(outTable) %in% colnames(best)]<-best[,colnames(best) %in% colnames(outTable)]

bestModels[21,]<-c("TXFW","TXCC",Model="AM2N2mG")
```


## Summarizing output for best models

```{r dadiSummaryPlot,fig.width=6,fig.height=6}
pop.list<-c(unique(bestModels[,1]),"TXCC")
colors<-c(SI="gray66",IM="#1f78b4",AM="#b2df8a",SC="#a6cee3")
lbcols<-c("#2166ac",'#1b7837',
          "#2166ac",'#af8dc3',"#2166ac",
          "#2166ac",'#762a83')
par(mfrow=c(length(pop.list),length(pop.list)),mar=c(0,0,0,0),oma=c(3,0,3,0))
for(i in 1:length(pop.list)){
  for(j in 1:length(pop.list)){
    
    exists<-nrow(bestModels[which(bestModels[,1] %in% pop.list[i] & 
                                    bestModels[,2] %in% pop.list[j]),])
    if(exists>0){
      
      model<-bestModels[which(bestModels[,1] %in% pop.list[i] &
                              bestModels[,2] %in% pop.list[j]),"Model"]
      col<-colors[names(colors) %in% gsub("(\\w\\w).*","\\1",model)]
      plot(c(0,1),c(0,1),type='n',axes=FALSE,xlab="",ylab="",bty='n')
      rect(0,0,1,1,col=alpha(col,0.75),border=col)
      legend("center",model,cex=2,bg='transparent',bty='n',
             adj=0.1,xjust=0,text.font = 2)
    } else{
      plot(c(0,1),c(0,1),type='n',axes=FALSE,xlab="",ylab="",bty='n')
    }
    if(i==1){
      mtext(pop.list[j],3,font = 2,outer = FALSE,cex=1.5,col=lbcols[j])
    }
    if(j==1){
      mtext(pop.list[i],2,font = 2,outer = FALSE,line=-1.6,
            cex=1.5,xpd=TRUE,col=lbcols[i])
    }
  }
}

```

```{r showBestTable}
landscape(kable(outTable,booktabs=TRUE))  %>%
  kable_styling(latex_options="scale_down")
```


## References



