---
title: Supplemental Material for "Multiple colonizations of freshwater by the Gulf pipefish reveal a shared genomic signature of adaptation"
preprint: false
author: 
  - name: Sarah P. Flanagan
    affiliation: 1
    corresponding: true
    email: spflanagan.phd@gmail.com
  - name: Emily Rose
    affiliation: 2
  - name: Adam Jones
    affiliation: 3
affiliation:
  - code: 1
    address: School of Biological Sciences, University of Canterbury, 4800 Private Bag, Christchurch 8140 New Zealand
  - code: 2
    address: Department of Biology, The University of Tampa, Tampa, FL 33606 USA
  - code: 3
    address: Department of Biological Sciences, University of Idaho, Moscow, ID 83844 USA
abstract: >
  This document includes supplementary material for the paper. 
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
bibliography: programs.bib
output:
  pdf_document:
    fig_caption: yes
    keep_tex: yes
    number_sections: no
    toc: true
    toc_depth: 2
    template: manuscript.latex
  html_document: null
  word_document: null
fontsize: 11pt
capsize: normalsize
csl: molecular-ecology.csl
documentclass: article
spacing: singlespacing
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,out.extra='',warning=FALSE,message=FALSE)
knitr::opts_knit$set(root.dir='../fwsw_results/')
```
```{r source}
source("../../gwscaR/R/gwscaR.R")
source("../../gwscaR/R/gwscaR_plot.R")
source("../../gwscaR/R/gwscaR_utility.R")
source("../../gwscaR/R/gwscaR_fsts.R")
source("../../gwscaR/R/gwscaR_popgen.R")
source("../../gwscaR/R/vcf2dadi.R")
source("../R/250_dadi_analysis.R")
library(knitr)
pop.list<-c("ALFW","ALST","FLCC","FLLG","LAFW","TXCC","TXFW")
```

Some useful functions are in the `250_dadi_analysis.R` script.

# Create a dadi SNPs file from the vcf {-}

```{r create_dadi, eval=FALSE}
vcf<-parse.vcf("stacks/populations_subset75/batch_2.vcf")
dadi<-vcf2dadiSNPs(vcf,pop.list = pop.list,filename = "dadi_analysis/fwsw.dadi.snps")
dadi<-read.delim("dadi_analysis/fwsw.dadi.snps")
projections<-unlist(lapply(pop.list,function(pop){ 
  n<-length(grep(pop,colnames(vcf)))
}))*2


names(projections)<-pop.list
best.pops<-c("TXFW","TXCB","LAFW","ALST","ALFW","FLLG","FLCC",
             "TXFW ","TXCB ","LAFW ","ALST ","ALFW ","FLLG ","FLCC ")
best.pops<-c("TXFW","TXCB","LAFW","ALST","ALFW","FLLG","FLCC",
             "TXFW.1","TXCB.1","LAFW.1","ALST.1","ALFW.1","FLLG.1","FLCC.1")
dadi<-dadi[,c("Ingroup","Outgroup","Allele1",best.pops[1:7],
              "Allele2",best.pops[8:14],"GeneID","Position")]
dim(dadi[rowSums(dadi[,colnames(dadi) %in% best.pops])==
           sum(projections[best.pops[1:7]]),])

#transform this and remove everything but population info for ease of operations
dd<-as.data.frame(cbind(colnames(dadi)[colnames(dadi) %in% best.pops],
                        t(dadi[,colnames(dadi) %in% best.pops])),
                  row.names = 0,stringsAsFactors = FALSE)
#rename the pops to be factors, essentially
dd$V1<-gsub("\\.1","",dd$V1)
#calculate sums per locus for each population
ns<-data.frame(do.call(rbind,lapply(2:ncol(dd),function(c){ 
  tapply(dd[,c],dd$V1,function(x){ sum(as.numeric(x)) }) })))
rownames(ns)<-dadi$GeneID
nozeros<-ns[!(apply(ns, 1, function(y) any(y == 0))),] #512 had zeroes
props<-apply(nozeros,2,function(x) { p<-x/max(x) } )
keepers<-props[apply(props, 1, function(y) any(y >=0.75)),]

#now save these in the dadi format
write.table(dadi[dadi$GeneID %in% rownames(keepers),],
            "dadi_analysis/fwsw.dadi.pruned.snps",
            quote=FALSE,row.names = FALSE,
            col.names = c("Ingroup","Outgroup","Allele1",
                          "TXFW","TXCB","LAFW","ALST","ALFW","FLLG","FLCC",
                          "Allele2","TXFW","TXCB","LAFW","ALST","ALFW",
                          "FLLG","FLCC","GeneID","Position"))
```

```{r check_dadis, eval=FALSE}
vcf.files<-c("stacks/populations_subset50/batch_2.vcf",
             "stacks/populations_subset75/batch_2.vcf",
             #"stacks/populations_subset85/batch_2.vcf",
             #"stacks/populations_subset90/batch_2.vcf",
             #"stacks/populations_subset95/batch_2.vcf",
             "stacks/populations_subset100/batch_2.vcf")

dadis<-lapply(vcf.files,function(file){
  full.vcf<-parse.vcf(file)
  vcf<-choose.one.snp(full.vcf)
  vcf.out.name<-gsub("batch_2.vcf","batch_2.pruned.vcf",file)
  write.table(x = vcf,file = vcf.out.name,quote = FALSE,
              sep='\t',row.names = FALSE,col.names = TRUE)
  dadi.out<-paste("dadi_analysis/fwsw",
                  gsub(".*populations_subset(\\d+).*","\\1",file),
                  ".dadi.snps",sep="")
  pop.list<-c("ALFW","ALST","FLCC","FLLG","LAFW","TXCC","TXFW")
  dadi<-vcf2dadiSNPs(vcf,pop.list = pop.list,filename = dadi.out)
  return(dadi)
})

#compare
kable(cbind(file=vcf.files,numer_snps=unlist(lapply(dadis,nrow))))
```

TODO: Write code to read in the files and compare

I'm going to use the 75% of individuals dataset.

## Running dadi {-}

I ran the optimize functions from (dportik's $\delta_A\delta_I$ pipeline)[https://github.com/dportik/dadi_pipeline] for each pairwise combination of the 7 populations with the 26 models from Rougeux. These ran using python2.7 (because I used $\delta_A\delta_I$ v. xxx) using individual scripts for each population pair and each model (in dadi_scripts/). These scripts were generated using `252_generate_dadi_scripts.sh`. Each of the population-pair-model scripts runs `252_pairwise_dadi.py`, which requires the dadi-pipeline scripts and the file `rougeux_models.py`, which contains the 26 models run here. Each script was run using either `252_run_dadi_scripts.sh` or `run_dadi_scripts_abacus.sh`, depending on whether the scripts were run on a local machine (C001KR or rccuser\@10.195.0.46) or on the cluster (abacus).

This is a visualization of all of the different models.

```{r dadiModels,fig.height=10,fig.width=8,dpi=300,dev='png',fig.path="../figs/"}
figPath<-"../figs/dadi_models/"
mods<-c("SI","IM","AM","SC")
adds<-c("","G","2N","2NG","2m","2mG","2N2m","2N2mG")

par(mfrow=c(8,4),oma=c(1,6,2,1),mar=c(1,1,1,1))

for(j in 1:length(adds)){
  for(i in 1:length(mods)){
    plot(c(0,1),c(0,1),type='n',axes=FALSE)
    imgName<-paste0(figPath,mods[i],adds[j],".PNG")
    if(file.exists(imgName)){
      img<-readPNG(imgName)
      rasterImage(img,0,0,1,1,xpd=TRUE)
    }
  }
}

# add headers
mtext("Strict Isolation",side=3,outer=TRUE,at=0.125,font=2)
mtext("Isolation w/ Migration",side=3,outer=TRUE,at=0.375,font=2)
mtext("Ancient Migration",side=3,outer=TRUE,at=0.62,font=2)
mtext("Secondary Contact",side=3,outer=TRUE,at=0.875,font=2)

# add rownames
mtext("Growth",side=2,outer=TRUE,at=(7/8)-0.05,font=2,las=1,adj=0.65)
mtext("Introgression",side=2,outer=TRUE,at=(6/8)-0.05,font=2,las=1,adj=0.65)
mtext("Introgression+
Growth",side=2,outer=TRUE,at=(5/8)-0.05,font=2,las=1,adj=0.65)
mtext("Het. Migration",side=2,outer=TRUE,at=(4/8)-0.05,font=2,las=1,adj=0.65)
mtext("Het. Migration+
Growth",side=2,outer=TRUE,at=(3/8)-0.05,font=2,las=1,adj=0.65)
mtext("Introgression+
Het. Migration",side=2,outer=TRUE,at=(2/8)-0.05,font=2,las=1,adj=0.65)
mtext("Introgression+
Het. Migration+
Growth",side=2,outer=TRUE,at=(1/8)-0.05,font=2,las=1,adj=0.65)

```


Single population plots were generated using `250_dadiPlots.py`.

To run the models I saved all of the population-pair-model combinations to a file to make it easier to run specific combinations.


```{r modelCombos,eval=FALSE}
pops=c('FLLG', 'FLCC', 'ALFW', 'ALST', 'LAFW', 'TXFW', 'TXCC')   #
models=c('SI', 'IM', 'AM', 'SC', 'SI2N', 'SIG', 'SI2NG', 'IMG', 
         'IM2N', 'IM2m', 'IM2NG', 'IM2mG', 'AM2N', 'AMG', 'AM2m', 
         'AM2NG', 'AM2N2m', 'AM2mG', 'AM2N2mG', 'SCG' ,'SC2N', 
         'SC2m', 'SC2NG', 'SC2N2m' ,'SC2mG' ,'SC2N2mG')
 



############ CREATE POP COMBOS TO RUN ############
tasks=NULL
for(i in 1:(length(pops)-1)){
  for(j in (i+1):length(pops)){
    for(mod in 1:length(models)){
      tasks<-c(tasks,paste(pops[i],pops[j],models[mod],sep="_"))
    }
  }
}

write.csv(tasks,"model_list_dadi.csv",row.names = TRUE,quote=FALSE)
```



## Analyzing dadi output {-}

```{r dadi_InitialCompare}
dadi_initialCompare<-function(comparison){
  opts<-dadi.modelcomp(path = paste0("dadi_results/",comparison,"/"),
                               pattern="*optimized.*",id=comparison)
  comp<-modelComparison(opts)
  Ns<-tapply(opts$log.likelihood,opts$Model,length)
  plot(as.numeric(opts$log.likelihood)~
              as.factor(opts$Model),las=3,
       ylim=c(min(as.numeric(opts$log.likelihood)-50),
              max(as.numeric(opts$log.likelihood)+50)),
        pch=19,xlab="",ylab="Maximum log likelihood")
  text(x=1:length(Ns),
       y=rep(min(as.numeric(opts$log.likelihood)-50),length(Ns)),
       Ns,cex=0.75)
  return(list(opts,comp))
}
```


### FLLG vs FLCC


```{r fllgOpts, fig.cap="Comparison of log likelihoods for FLLG-FLCC models"}
fllg_flcc<-dadi_initialCompare("FLLG_FLCC")
```


It's looking like IM2mG model has a higher on average log likelihood than SC2mG. Let's see if the IM2mG parameter distributions are better

```{r fllgBestMods}
flcc_fllg_sc2mg<-do.call(rbind,
                         extract_dadi_params(fllg_flcc[[1]][fllg_flcc[[1]]$Model==
                                                              "SC2mG",]))
flcc_fllg_im2mg<-do.call(rbind,
                         extract_dadi_params(fllg_flcc[[1]][fllg_flcc[[1]]$Model==
                                                              "IM2mG",]))
```

```{r fllgCompareMods}
par(mfrow=c(1,2))
boxplot(flcc_fllg_sc2mg[,11:22])
points(1:12,
       flcc_fllg_sc2mg[which.max(flcc_fllg_sc2mg$log.likelihood),11:22],
       pch=8,col="red",lwd=2)

boxplot(flcc_fllg_im2mg[,11:21])
points(1:11,
       flcc_fllg_im2mg[which.max(flcc_fllg_im2mg$log.likelihood),11:21],
       pch=8,col="red",lwd=2)
```

### FLLG vs ALFW
```{r saveBest1}
bestModels<-data.frame(Pop1="FLFW",
                       Pop2="FLCC",
                       Model="IM2mG",
                       stringsAsFactors = FALSE)
```

```{r ffafOpts, fig.cap="Comparison of log likelihoods for FLLG-ALFW models" }
fllg_alfw<-dadi_initialCompare("FLLG_ALFW")
```
```{r saveBest2}
bestModels[2,]<-c("FLFW","ALFW",Model="AM2NG")
```


### FLLG vs ALST

```{r ffasOpts, fig.cap="Comparison of log likelihoods for FLLG-ALST models"}
fllg_alst<-dadi_initialCompare("FLLG_ALST")
```

```{r saveBest3}
bestModels[3,]<-c("FLFW","ALST",Model="AM2mG")
```

### FLLG vs LAFW

```{r fflfOpts, fig.cap="Comparison of log likelihoods for FLLG-LAFW models"}
fllg_lafw<-dadi_initialCompare("FLLG_LAFW")
```
```{r saveBest4}
bestModels[4,]<-c("FLFW","LAFW",Model="SC2N2mG")
```

### FLLG vs TXFW

```{r fftfOpts, fig.cap="Comparison of log likelihoods for FLLG-TXFW models"}
fllg_txfw<-dadi_initialCompare("FLLG_TXFW")
```
```{r saveBest5}
bestModels[5,]<-c("FLFW","TXFW",Model="AM2N2mG")
```

### FLLG vs TXCC

```{r fftsOpts, fig.cap="Comparison of log likelihoods for FLLG-TXCC models"}
fllg_txcc<-dadi_initialCompare("FLLG_TXCC")
```
```{r saveBest6}
bestModels[6,]<-c("FLFW","TXCC",Model="AM2mG")
```

### FLCC vs ALFW

```{r fsafOpts, fig.cap="Comparison of log likelihoods for FLCC-ALFW models"}
flcc_alfw<-dadi_initialCompare("FLCC_ALFW")
```
```{r saveBest7}
bestModels[7,]<-c("FLCC","ALFW",Model="IM")
```

### FLCC vs ALST

```{r fsasOpts, fig.cap="Comparison of log likelihoods for FLCC-ALST models"}
flcc_alst<-dadi_initialCompare("FLCC_ALST")
```
```{r saveBest8}
bestModels[8,]<-c("FLCC","ALST",Model="SC2N")
```

### FLCC vs LAFW

```{r fslfOpts, fig.cap="Comparison of log likelihoods for FLCC-LAFW models"}
flcc_lafw<-dadi_initialCompare("FLCC_LAFW")
```
```{r saveBest9}
bestModels[9,]<-c("FLCC","LAFW",Model="IM2mG")
```

### FLCC vs TXFW

```{r fstfOpts, fig.cap="Comparison of log likelihoods for FLCC-TXFW models"}
flcc_txfw<-dadi_initialCompare("FLCC_TXFW")
```
```{r saveBest10}
bestModels[10,]<-c("FLCC","TXFW",Model="IM2mG")
```

### FLCC vs TXCC

```{r fstsOpts, fig.cap="Comparison of log likelihoods for FLCC-TXCC models"}
flcc_txcc<-dadi_initialCompare("FLCC_TXCC")
```

```{r saveBest11}
bestModels[11,]<-c("FLCC","TXCC",Model="IM2mG")
```

### ALFW vs ALST

```{r afasOpts, fig.cap="Comparison of log likelihoods for ALFW-ALST models"}
alfw_alst<-dadi_initialCompare("ALFW_ALST")
```

```{r saveBest12}
bestModels[12,]<-c("ALFW","ALST",Model="SC2m")
```

### ALFW vs LAFW

```{r aflfOpts, fig.cap="Comparison of log likelihoods for ALFW-LAFW models"}
alfw_lafw<-dadi_initialCompare("ALFW_LAFW")
```

```{r saveBest13}
bestModels[13,]<-c("ALFW","LAFW",Model="SC2N2mG")
```

### ALFW vs TXFW

```{r aftfOpts, fig.cap="Comparison of log likelihoods for ALFW-TXFW models"}
alfw_txfw<-dadi_initialCompare("ALFW_TXFW")
```

```{r saveBest14}
bestModels[14,]<-c("ALFW","TXFW",Model="IM2N")
```


### ALFW vs TXCC

```{r aftsOpts, fig.cap="Comparison of log likelihoods for ALFW-TXCC models"}
alfw_txcc<-dadi_initialCompare("ALFW_TXCC")
```

```{r saveBest15}
bestModels[15,]<-c("ALFW","TXCC",Model="SC2N2mG")
```


### ALST vs LAFW

```{r aslfOpts, fig.cap="Comparison of log likelihoods for ALST-LAFW models"}
alst_lafw<-dadi_initialCompare("ALST_LAFW")
```
```{r saveBest16}
bestModels[16,]<-c("ALST","LAFW",Model="IM2NG")
```


### ALST vs TXFW

```{r astfOpts, fig.cap="Comparison of log likelihoods for ALST-TXFW models"}
alst_txfw<-dadi_initialCompare("ALST_TXFW")
```
```{r saveBest17}
bestModels[17,]<-c("ALST","TXFW",Model="IM")
```


### ALST vs TXCC

```{r astsOpts, fig.cap="Comparison of log likelihoods for ALST-TXCC models"}
alst_txc<-dadi_initialCompare("ALST_TXCC")
```
```{r saveBest18}
bestModels[18,]<-c("ALST","TXCC",Model="AM2mG")
```


### LAFW vs TXFW

```{r lftfOpts, fig.cap="Comparison of log likelihoods for LAFW-TXFW models"}
lafw_txfw<-dadi_initialCompare("LAFW_TXFW")
```
```{r saveBest19}
bestModels[19,]<-c("LAFW","TXFW",Model="SCG")
```


### LAFW vs TXCC

```{r lftsOpts, fig.cap="Comparison of log likelihoods for LAFW-TXCC models"}
lafw_txcc<-dadi_initialCompare("LAFW_TXCC")
```
```{r saveBest20}
bestModels[20,]<-c("LAFW","TXCC",Model="SC2NG")
```


### TXFW vs TXCC

```{r tftsOpts, fig.cap="Comparison of log likelihoods for TXFW-TXCC models"}
txfw_txcc<-dadi_initialCompare("TXFW_TXCC")
```
```{r saveBest21}
bestModels[21,]<-c("TXFW","TXCC",Model="AM2N2mG")
```

