---
title: Supplemental Material \# 2 for "Multiple colonizations of freshwater by the Gulf pipefish reveal a shared genomic signature of adaptation"
preprint: false
author: 
  - name: Sarah P. Flanagan
    affilnum: 1
    corresponding: true
    email: spflanagan.phd@gmail.com
  - name: Emily Rose
    affilnum: 2
  - name: Adam Jones
    affilnum: 3
affiliation:
  - affilnum: 1
    affil: School of Biological Sciences, University of Canterbury, 4800 Private Bag, Christchurch 8140 New Zealand
  - affilnum: 2
    affil: Department of Biology, The University of Tampa, Tampa, FL 33606 USA
  - affilnum: 3
    affil: Department of Biological Sciences, University of Idaho, Moscow, ID 83844 USA
abstract: >
  This document includes supplementary material for the paper, showing expanded methods and results from the diffusion approximations.
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
bibliography: programs.bib
output:
  bookdown::pdf_document2:
    fig_caption: yes
    keep_tex: yes
    number_sections: no
    toc: true
    toc_depth: 2
    template: manuscript.latex
  html_document: null
  word_document: null
fontsize: 11pt
capsize: normalsize
csl: molecular-ecology.csl
documentclass: article
spacing: singlespacing
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,out.extra='',fig.pos="H",
                      warning=FALSE,message=FALSE)
knitr::opts_knit$set(root.dir='../fwsw_results/')
```
```{r source}
source("../../gwscaR/R/gwscaR.R")
source("../../gwscaR/R/gwscaR_plot.R")
source("../../gwscaR/R/gwscaR_utility.R")
source("../../gwscaR/R/gwscaR_fsts.R")
source("../../gwscaR/R/gwscaR_popgen.R")
source("../../gwscaR/R/vcf2dadi.R")
source("../R/250_dadi_analysis.R")
library(knitr); library(kableExtra)
library(png)
library(scales)
pop.list<-c("ALFW","ALST","FLCC","FLLG","LAFW","TXCC","TXFW")
fw_pops<-c("TXFW","LAFW","ALFW","FLLG")
sw_pops<-c("TXCC","ALST","ALST","FLCC")
```



# Create a dadi SNPs file from the vcf {-}


First I am filtering the full dataset, which has 540066 SNPs at 111391 RAD loci. First I'll choose a single SNP per locus (so I have 111391 SNPs), and then I'll filter to minimize missingness by keeping SNPs 

```{r filterLowCovInds,eval=FALSE}
vcf<-parse.vcf("stacks/batch_2.vcf")
# choose one SNP per locus
vcf1<-choose.one.snp(vcf)
rownames(vcf1)<-NULL
# remove any indivdiuals with > 75% missingness
ind_cov<-apply(vcf1[,10:ncol(vcf1)],2,function(x){ return(length(grep("\\.\\/\\.",x))/length(x)) })
ind_rem<-which(ind_cov>=0.55)
vcf2<-vcf1[,!(colnames(vcf1) %in% names(ind_rem))]
write.table(vcf2,"dadi_results/filtered.vcf",col.names = TRUE,row.names = FALSE,quote=FALSE,sep='\t')
```

```{r create_dadi, eval=FALSE}
# ID SNPs with high coverage (low missingness) in each population
keep_snps<-lapply(pop.list,function(pop,vcf){
  pop_vcf<-vcf[,c(1:9,grep(pop,colnames(vcf)))]
  pop_prop<-apply(pop_vcf[,10:ncol(pop_vcf)],1,function(x){ return(length(grep("\\.\\/\\.",x))/length(x)) })
  return(pop_vcf$ID[which(pop_prop<=0.1)])
},vcf=vcf2)
names(keep_snps)<-pop.list

pop_vcfs<-list()
for(i in 1:length(fw_pops)){
  keep_ids<-keep_snps[[fw_pops[i]]][keep_snps[[fw_pops[i]]] %in% keep_snps[[sw_pops[i]]]]
  pop_vcfs[[i]]<-vcf2[which(vcf2$ID %in% keep_ids),c(1:9,grep(fw_pops[i],colnames(vcf2)),grep(sw_pops[i],colnames(vcf2)))]
  pop_dadi<-vcf2dadiSNPs(vcf=pop_vcfs[[i]],pop.list=c(fw_pops[i],sw_pops[i]),filename=paste0("dadi_results/",fw_pops[i],"_",sw_pops[i],".dadi.snps"))
}
names(pop_vcfs)<-paste0(fw_pops,"_",sw_pops)

projections<-unlist(lapply(pop.list,function(pop){ 
  n<-length(grep(pop,colnames(vcf2)))
}))*2

```


```{r comparePopPairs}
files<-list.files(pattern="dadi.snps",path="dadi_results",full.names = TRUE)
projections<-c(TXFW=54,LAFW=52,ALFW=60,FLLG=76,TXCC=86,ALST=62,FLCC=72)

dadis<-lapply(files,read.delim)
dadiSum<-data.frame(PopPair=gsub("dadi_results/(\\w+_\\w+).dadi.snps","\\1",files),
                    Pop1=projections[gsub("dadi_results/(\\w+)_(\\w+).dadi.snps","\\1",files)]/2,
                    Pop2=projections[gsub("dadi_results/(\\w+)_(\\w+).dadi.snps","\\2",files)]/2,
                    NumberSnps=do.call(rbind,lapply(dadis,nrow)))
#compare
kable(dadiSum,"latex",booktabs=TRUE,
      col.names = c("Population Pair","Number of Individuals in Pop 1","Number of Individuals in Pop 2","Number of SNPs"),
      caption = "For each analysis of the freshwater-saltwater pairs, the retained number of individuals (the values in this table were multiplied by 2 for the projections used in dadi) in each population and the number of SNPs retained in each analysis.")%>%
  kable_styling(latex_options=c("HOLD_position", "scale_down"))
```


Each pairwise population comparison used a slightly different set of SNPs (Table \@ref(tab:compareSNPs)) to minimise missingness and allow for the inclusion of SNPs that were singletons and without a minor allele frequency cutoff.

## Running dadi {-}

I ran the optimize functions from (dportik's $\delta_A\delta_I$ pipeline)[https://github.com/dportik/dadi_pipeline] for each pairwise combination of the 7 populations with the 26 models from [@rougeuxModelingMultipleFacets2017]. These ran using python2.7 and $\delta_A\delta_I$ v. 1.7.0, using individual scripts for each population pair and each model (in dadi_scripts/). These scripts were generated using `252_generate_dadi_scripts.sh`. Each of the population-pair-model scripts runs `252_pairwise_dadi.py`, which requires the dadi-pipeline scripts and the file `rougeux_models.py`, which contains the 26 models run here. Each script was run using either `252_run_dadi_scripts.sh` or `run_dadi_scripts_abacus.sh`, depending on whether the scripts were run on a local machine (C001KR or rccuser\@10.195.0.46) or on the cluster (abacus).


```{r dadiModels,fig.height=8,fig.width=6,dpi=300,dev='png',fig.path="../figs/", fig.cap="Schematic diagrams of each model, color-coded by base model (Strict Isolation, Isolation with Migration, Ancient Migration, or Secondary Contact). 'Linked Sel.' refers to linked selection, where a sub-section of the genome experienced a different effective population size than the remainder of the genome (hrf). 'Het. Migration' refers to differential introgression, where some regions of the genome experienced different migration rates than others. Each row demonstrates how parameters were added to the base models. Strict Isolation models contained now migration, so only linked selection and growth could be incorporated. Under Isolation with Migration scenarios, differentiating between linked selection and heterogeneous migration is difficult, so the models that combined both were not included for the Isolation with Migration scenarios."}
figPath<-"../figs/dadi_models/"
mods<-c("SI","IM","AM","SC")
adds<-c("","G","2N","2NG","2m","2mG","2N2m","2N2mG")

par(mfrow=c(8,4),oma=c(1,6,4,1),mar=c(0,0,0,0))

for(j in 1:length(adds)){
  for(i in 1:length(mods)){
    plot(c(0,1),c(0,1),type='n',axes=FALSE)
    imgName<-paste0(figPath,mods[i],adds[j],".PNG")
    if(file.exists(imgName)){
      img<-readPNG(imgName)
      rasterImage(img,0,0,1,1,xpd=TRUE)
    }
  }
}

# add headers
mtext("Strict
Isolation",side=3,outer=TRUE,at=0.125,font=2)
mtext("Isolation w/ 
Migration",side=3,outer=TRUE,at=0.375,font=2)
mtext("Ancient
Migration",side=3,outer=TRUE,at=0.62,font=2)
mtext("Secondary
Contact",side=3,outer=TRUE,at=0.875,font=2)

# add rownames
mtext("Growth",side=2,outer=TRUE,at=(7/8)-0.05,font=2,las=1,adj=0.65)
mtext("Linked Sel.",side=2,outer=TRUE,at=(6/8)-0.05,font=2,las=1,adj=0.65)
mtext("Linked Sel.+
Growth",side=2,outer=TRUE,at=(5/8)-0.05,font=2,las=1,adj=0.65)
mtext("Het. Migration",side=2,outer=TRUE,at=(4/8)-0.05,font=2,las=1,adj=0.65)
mtext("Het. Migration+
Growth",side=2,outer=TRUE,at=(3/8)-0.05,font=2,las=1,adj=0.65)
mtext("Linked Sel.+
Het. Migration",side=2,outer=TRUE,at=(2/8)-0.05,font=2,las=1,adj=0.65)
mtext("Linked Sel.+
Het. Migration+
Growth",side=2,outer=TRUE,at=(1/8)-0.05,font=2,las=1,adj=0.65)

```


Single population plots were generated using `250_dadiPlots.py`.

To run the models I saved all of the population-pair-model combinations to a file to make it easier to run specific combinations.


```{r modelCombos,eval=FALSE}
fw_pops=c('TXFW', 'LAFW', 'ALFW', 'FLLG')
sw_pops=c('TXCC', 'ALST', 'ALST', 'FLCC')
models=c('SI', 'IM', 'AM', 'SC', 'SI2N', 'SIG', 'SI2NG', 'IMG', 
         'IM2N', 'IM2m', 'IM2NG', 'IM2mG', 'AM2N', 'AMG', 'AM2m', 
         'AM2NG', 'AM2N2m', 'AM2mG', 'AM2N2mG', 'SCG' ,'SC2N', 
         'SC2m', 'SC2NG', 'SC2N2m' ,'SC2mG' ,'SC2N2mG')
 



############ CREATE POP COMBOS TO RUN ############
tasks=NULL
for(i in 1:(length(fw_pops))){
  for(mod in 1:length(models)){
    tasks<-c(tasks,paste(fw_pops[i],sw_pops[i],models[mod],sep="_"))
  }
}
tasks<-data.frame(tasks,0:(length(tasks)-1))
write.csv(tasks,"model_list_dadi_Aug2020.csv",row.names = FALSE,col.names = FALSE,quote=FALSE)
```

```{r dadiParamsTable}
paramsMeanings<-data.frame(parameter=c("theta","b1","b2","m12","m21","me12","me21",
                                       "nu1","nu2","P","Ts","Tsc","hrf",
                                       "Q","bf","nr","Tam"),
                           description=c("ancestral population size",
                                         "population growth coefficient for population 1",
                                         "population growth coefficient for population 2", 
                                         "migration rate from population 2 to population 1",
                                         "migration from population 1 to population 2", 
                                         "effective migration from population 2 to population 1 in genomic islands",
                                         "effective migration from population 1 to population 2 in genomic islands",
                                         "Size of population 1 after the split",
                                         "Size of population 2 after the split",
                                         "The proportion of the genome evolving neutrally",
                                         "The scaled time between the split and the present/secondary contact",
                                         "The scaled time between secondary contact and the present",
                                         "The degree to which Ne is locally reduced due to background selection and selective sweeps",
                                         "The proportion of the genome with reduced Ne due to selection at linked sites",
                                         "The extent of reduced Ne in non-recombining regions",
                                         "Proportion of non- or low-recombining regions",
                                         "The scaled time between ancient migration and the present"),
                           models=c("All",
                                    "All models with growth (G)",
                                    "All models with growth (G)",
                                    "Models with migration (SC, AM, and IM)",
                                    "Models with migration (SC, AM, and IM)",
                                    "Models with heterogeneous migration (with 2m)",
                                    "Models with heterogeneous migration (with 2m)",
                                    "All",
                                    "All",
                                    "Models with heterogeneous migration (with 2m)",
                                    "All",
                                    "Models with secondary contact (SC)",
                                    "Models with background selection (with 2N)",
                                    "Models with background selection (with 2N)",
                                    "Strict isolation and background selection (SI2N)",
                                    "Strict isolation and background selection (SI2N)",
                                    "Models with ancestral migratino (AM)"))

kable(paramsMeanings,"latex",booktabs=TRUE,longtable=TRUE,
      caption="Table describing all of the parameters from the dadi models and which models they appear in.") %>%
  kable_styling(latex_options=c("HOLD_position","repeat_header")) %>%
  column_spec(2,width="17.5em") %>%
  column_spec(3,width="17.5em")
```



## Analyzing dadi output {-}

```{r dadi_InitialCompare}
dadi_initialCompare<-function(comparison,opt_pattern="*optimized.*",modSel="max"){
  opts<-dadi.modelcomp(path = paste0("dadi_results/",comparison,"/"),
                               pattern=opt_pattern,id=comparison,last=FALSE)
  comp<-modelComparison(opts,modSel)
  Ns<-tapply(opts$log.likelihood,opts$Model,length)
  plot(as.numeric(opts$log.likelihood)~
              as.factor(opts$Model),las=3,
       ylim=c(min(as.numeric(opts$log.likelihood)-50),
              max(as.numeric(opts$log.likelihood)+50)),
        pch=19,xlab="",ylab="log likelihood")
  text(x=1:length(Ns),
       y=rep(min(as.numeric(opts$log.likelihood)-50),length(Ns)),
       Ns,cex=0.75)
  return(list(opts,comp))
}
```
```{r}
dadi_pairSummary<-function(compareList){
  sumDat<-extract_dadi_params(compareList[[2]])
  keepnames<-unique(unlist(lapply(sumDat,colnames)))
  keepnames<-keepnames[!keepnames %in% c("Replicate","params","optimized_params","rank","id")]
  keepnames<-c("PopPair",keepnames)
  sumTab<-data.frame(matrix(nrow = length(sumDat),ncol=length(keepnames),
                              dimnames = list(NULL,keepnames)))
  for(i in 1:length(sumDat)){
    sumTab[i,1]<-sumDat[[i]]$id
    
    sumTab[i,colnames(sumTab)%in%colnames(sumDat[[i]])]<-sumDat[[i]][,colnames(sumDat[[i]])%in%colnames(sumTab)]
  }
  return(sumTab)
}

```


To compare the $\delta_A\delta_I$ results for each population pair, we used custom functions (in the `250_dadi_analysis.R` script) to choose the run of each model with the highest median log likelihood. We opted to use the run with the median log likelihood because we observed some of our models with large log likelihoods explored paramter space not representative of all of the runs of that particular model and population pair (for example, see Figure \@ref(fig:fllgCompareMods)). We then compared log likelihoods, AIC scores, and model scores to select the best model to desribe the demographic history of each population pair. 

### FLFW vs FLCC


```{r fllgOptsSimple, fig.cap="Comparison of log likelihoods for FLFW-FLCC models"}
fllg_flcc<-dadi_initialCompare("FLLG_FLCC",opt_pattern = "\\.\\w\\w.optimized.txt",modSel="median")
```


```{r fffwTableSimple}
sumTab<-dadi_pairSummary(fllg_flcc)

landscape(kable(sumTab,"latex",booktabs=TRUE,
      caption = "Table showing the results of the run with the median log likelihood for each model for the FLFW vs FLCC comparison. The parameters are not converted to biologically realistic values, and their meanings are presented in Table S2.2.",
      row.names = FALSE)) %>%
  kable_styling(latex_options=c("scale_down","HOLD_position"))

best_mod<-sumTab$Model[which.max(sumTab$model_score)]

```


The model with the highest model score was `r best_mod`, with SC2mG having the second-best model score (Figre \@ref(fig:fllgOpts), Table \@ref(tab:fffwTable)). We compared the parameter estimates for these two models, visualizing the parameter estimates for each (Figure \@ref(fig:fllgCompareMods)). This figure demonstrates that the model replicate with the maximum log likelihood tended to have extreme estimates of the b1 parameter, reinforcing our choice to choose the best model runs based on model score. 

```{r fllgBestMods, eval=FALSE}
flcc_fllg_sc2mg<-do.call(rbind,
                         extract_dadi_params(fllg_flcc[[1]][fllg_flcc[[1]]$Model==
                                                              "SC2mG",]))
flcc_fllg_im2mg<-do.call(rbind,
                         extract_dadi_params(fllg_flcc[[1]][fllg_flcc[[1]]$Model==
                                                              "IM2mG",]))
```

```{r fllgCompareMods,fig.width=8, fig.cap="Boxplot showing the parameter estimates from the replicates of the SC2mG model (left) and the IM2mG model (right) for the FLFW vs FLCC pairwise comparison. The red stars represent the model parameters from the run with the highest log likelihood.", eval=FALSE}
par(mfrow=c(1,2))
boxplot(flcc_fllg_sc2mg[,11:22],main="SC2mG")
points(1:12,
       flcc_fllg_sc2mg[which.max(flcc_fllg_sc2mg$log.likelihood),11:22],
       pch=8,col="red",lwd=2)

boxplot(flcc_fllg_im2mg[,11:21],main="IM2mG")
points(1:11,
       flcc_fllg_im2mg[which.max(flcc_fllg_im2mg$log.likelihood),11:21],
       pch=8,col="red",lwd=2)
```
```{r setupOutTable}

outNames<-colnames(sumTab)

outTable<-data.frame(matrix(nrow = 1,ncol=length(outNames),
                            dimnames = list("",outNames)))
```



```{r saveBest1}
outTable[1,]<-sumTab[sumTab$Model==best_mod, colnames(outTable)]

bestModels<-data.frame(Pop1="FLFW",
                       Pop2="FLCC",
                       Model=best_mod,
                       stringsAsFactors = FALSE)
```

Based on these results, we conclude that the FLFW and FLCC populations -- which represent a freshwater population and its nearest saltwater neighbor -- have likely experienced recent migration (which is supported by both the IM2mG and the SC2mG models), with heterogeneous migration (i.e., differential introgression) across the genome, and changes in populations sizes.

```{r dadiPlots1,out.width="90%",fig.cap="Summary plot of the joint site frequency spectrum of the data for the FLFW vs FLCC comparison (left) and the best-fitting model (right), plus their residuals (bottom row)."}
multiname<-paste0("../figs/dadi/FLLG_FLCC_",best_mod,"multinom.png")
if(file.exists(multiname)){
  knitr::include_graphics(multiname)
}
```


### ALFW vs ALST

```{r afasOpts, fig.cap="Comparison of log likelihoods for ALFW-ALST models"}
alfw_alst<-dadi_initialCompare("ALFW_ALST",modSel="median")
```

```{r sumTab12}
sumTab<-dadi_pairSummary(alfw_alst)

landscape(kable(sumTab,"latex",booktabs=TRUE,
      caption = "Table showing the results of the run with the median log likelihood for each model for the ALFW vs ALST comparison. The parameters are not converted to biologically realistic values, and their meanings are presented in Table S2.2.",
      row.names = FALSE)) %>%
  kable_styling(latex_options=c("scale_down","HOLD_position"))
```
```{r BestMods12}
best_mod<-sumTab$Model[which.max(sumTab$model_score)]
alfw_alst_best<-do.call(rbind,
                      extract_dadi_params(
                        alfw_alst[[1]][alfw_alst[[1]]$Model==best_mod,]))
```
The model with the highest model score was `r best_mod`.
```{r saveBest12}

outTable[12,]<-sumTab[sumTab$Model==best_mod,colnames(outTable)]

bestModels[12,]<-c("ALFW","ALST",Model=best_mod)
```
```{r dadiPlots12,out.width="90%",fig.cap="Summary plot of the joint site frequency spectrum of the data for the ALFW vs ALST comparison (left) and the best-fitting model (right), plus their residuals (bottom row)."}
multiname<-paste0("../figs/dadi/ALFW_ALST_",best_mod,"multinom.png")
if(file.exists(multiname)){
  knitr::include_graphics(multiname)
}
```


### LAFW vs ALST

```{r aslfOpts, fig.cap="Comparison of log likelihoods for ALST-LAFW models"}
lafw_alst<-dadi_initialCompare("LAFW_ALST",modSel="median")
```

```{r sumTab16}
sumTab<-dadi_pairSummary(lafw_alst)

landscape(kable(sumTab,"latex",booktabs=TRUE,
      caption = "Table showing the results of the run with the median log likelihood for each model for the ALST vs LAFW comparison. The parameters are not converted to biologically realistic values, and their meanings are presented in Table S2.2.",
      row.names = FALSE)) %>%
  kable_styling(latex_options=c("scale_down","HOLD_position"))
```
```{r BestMods16}
best_mod<-sumTab$Model[which.max(sumTab$model_score)]
lafw_alst_best<-do.call(rbind,
                      extract_dadi_params(
                        lafw_alst[[1]][lafw_alst[[1]]$Model==best_mod,]))
```
The model with the highest model score was `r best_mod`.
```{r saveBest16}

outTable[16,]<-sumTab[sumTab$Model==best_mod,colnames(outTable)]

bestModels[16,]<-c("ALST","LAFW",Model=best_mod)
```
```{r dadiPlots16,out.width="90%",fig.cap="Summary plot of the joint site frequency spectrum of the data for the ALST vs LAFW comparison (left) and the best-fitting model (right), plus their residuals (bottom row)."}
multiname<-paste0("../figs/dadi/LAFW_ALST_",best_mod,"multinom.png")
if(file.exists(multiname)){
  knitr::include_graphics(multiname)
}
```


### TXFW vs TXCC

```{r tftsOpts, fig.cap="Comparison of log likelihoods for TXFW-TXCC models"}
txfw_txcc<-dadi_initialCompare("TXFW_TXCC",modSel="median")
```

```{r sumTab21}
sumTab<-dadi_pairSummary(txfw_txcc)

landscape(kable(sumTab,"latex",booktabs=TRUE,
      caption = "Table showing the results of the run with the median log likelihood for each model for the TXFW vs TXCC comparison. The parameters are not converted to biologically realistic values, and their meanings are presented in Table S2.2.",
      row.names = FALSE)) %>%
  kable_styling(latex_options=c("scale_down","HOLD_position"))
```
```{r BestMods21}
best_mod<-sumTab$Model[which.max(sumTab$model_score)]
txfw_txcc_best<-do.call(rbind,
                      extract_dadi_params(
                        txfw_txcc[[1]][txfw_txcc[[1]]$Model==best_mod,]))
```
The model with the highest model score was `r best_mod`.
```{r saveBest21, eval=FALSE}

outTable[21,]<-sumTab[sumTab$Model==best_mod,colnames(outTable)]

bestModels[21,]<-c("TXFW","TXCC",Model=best_mod)
```
```{r dadiPlots21,out.width="90%",fig.cap="Summary plot of the joint site frequency spectrum of the data for the TXFW vs TXCC comparison (left) and the best-fitting model (right), plus their residuals (bottom row)."}
multiname<-paste0("../figs/dadi/TXFW_TXCC_",best_mod,"multinom.png")
if(file.exists(multiname)){
  knitr::include_graphics(multiname)
}
```

## Summarizing output for best models

```{r dadiSummaryPlot,fig.width=8,fig.height=8, fig.cap="The best models for each population pair, color-coded by baseline demographic scenario (grey = strict isolation; green = ancestral migration; light blue = secondary contact; dark blue = isolation with migration). The abbreviation 2N refers to linked selection and 2m refers to heterogeneous migration/introgression, both of which are indicative of covariance among some genomic regions. The letter G indicates that the best model supports changes in population size since the population split." , eval=FALSE}
pop.list<-c(unique(bestModels[,1]),"TXCC")
colors<-c(SI="gray66",IM="#1f78b4",AM="#b2df8a",SC="#a6cee3")
lbcols<-c("#2166ac",'#1b7837',
          "#2166ac",'#af8dc3',"#2166ac",
          "#2166ac",'#762a83')
par(mfrow=c(length(pop.list),length(pop.list)),mar=c(0,0,0,0),oma=c(3,0,3,0))
for(i in 1:length(pop.list)){
  for(j in 1:length(pop.list)){
    
    exists<-nrow(bestModels[which(bestModels[,1] %in% pop.list[i] & 
                                    bestModels[,2] %in% pop.list[j]),])
    if(exists>0){
      
      model<-bestModels[which(bestModels[,1] %in% pop.list[i] &
                              bestModels[,2] %in% pop.list[j]),"Model"]
      col<-colors[names(colors) %in% gsub("(\\w\\w).*","\\1",model)]
      plot(c(0,1),c(0,1),type='n',axes=FALSE,xlab="",ylab="",bty='n')
      rect(0,0,1,1,col=alpha(col,0.75),border=col,lwd=5)
      legend("center",model,cex=1.5,bg='transparent',bty='n',
             adj=0.1,xjust=0,text.font = 2)
    } else{
      plot(c(0,1),c(0,1),type='n',axes=FALSE,xlab="",ylab="",bty='n')
    }
    if(i==1){
      mtext(pop.list[j],3,font = 2,outer = FALSE,cex=1.25,col=lbcols[j])
    }
    if(j==1){
      mtext(pop.list[i],2,font = 2,outer = FALSE,line=-1.6,
            cex=1.25,xpd=TRUE,col=lbcols[i])
    }
  }
}
write.csv(outTable,"dadi_outTable.csv",col.names = TRUE,row.names = FALSE,quote=FALSE)
```

The summary table will tell show all of the parameters for the best models. 

```{r showBestTable}
landscape(kable(outTable,"latex",booktabs=TRUE,
                caption="Table of all of the parameters for the best models for each population pair."))  %>%
  kable_styling(latex_options=c("scale_down","repeat_header","HOLD_position"))
```
```{r saveBestTable, eval=FALSE}
poporder<-c('TXFW_TXCC','ALST_LAFW','ALFW_ALST','FLLG_FLCC', 
            'LAFW_TXFW','ALFW_TXFW','FLLG_TXFW','ALFW_LAFW','FLLG_LAFW','FLLG_ALFW',
            'ALST_TXCC','FLCC_ALST','FLCC_TXCC',
            'ALST_TXFW','FLCC_TXFW',
            'LAFW_TXCC','FLCC_LAFW',
            'ALFW_TXCC','FLCC_ALFW',
            'FLLG_ALST','FLLG_TXCC')   
savTab<-outTable[match(poporder,outTable$PopPair),1:6]
savTab$PopPair<-gsub("FLLG","FLFW",savTab$PopPair)
write.csv(savTab,"dadi_summary.csv",col.names = TRUE,row.names = FALSE,quote=FALSE)

```


```{r paramsOrder, eval=FALSE}
params_order<-list(SI=c("nu1", "nu2", "Ts", "O"),
                    SI2N=c("nu1", "nu2", "Ts", "nr", "bf", "O"),
                    SIG=c("nu1", "nu2", "b1", "b2", "Ts", "O" ),
                    SI2NG=c("nu1", "nu2", "b1", "b2", "hrf", "Ts", "Q", "O"),
                    IM=c("nu1", "nu2", "m12", "m21", "Ts", "O" ),
                    IMG=c("nu1", "nu2", "b1", "b2", "m12", "m21", "Ts", "O"),
                    IM2N=c("nu1", "nu2", "hrf", "m12", "m21", "Ts", "Q", "O"),
                    IM2NG=c("nu1", "nu2", "b1", "b2", "hrf", "m12", "m21", "Ts", "Q", "O"),
                    IM2m=c("nu1", "nu2", "m12", "m21", "me12", "me21", "Ts", "P", "O"),
                    IM2mG=c("nu1", "nu2", "b1", "b2", "m12", "m21", "me12", "me21", "Ts", "P", "O"),
                    AM=c("nu1", "nu2", "m12", "m21", "Ts", "Tam", "O"),
                    AMG=c("nu1", "nu2", "b1", "b2", "m12", "m21", "Tam", "Ts", "O"),
                    AM2N=c( "nu1", "nu2", "hrf", "m12", "m21", "Tam", "Ts", "Q", "O"),
                    AM2N2m=c("nu1", "nu2", "hrf", "m12", "m21", "me12", "me21", "Tam", "Ts", "P", "Q", "O"),
                   AM2NG=c("nu1", "nu2", "b1", "b2", "hrf", "m12", "m21", "Tam", "Ts", "Q", "O"),
                   AM2m=c("nu1", "nu2", "m12", "m21", "me12", "me21", "Ts", "Tam", "P", "O"),
                   AM2mG=c( "nu1", "nu2", "b1", "b2", "m12", "m21", "me12", "me21", "Tam", "Ts", "P", "O"),
                   AM2N2mG=c("nu1", "nu2", "b1", "b2", "hrf", "m12", "m21", "me12", "me21", "Tam", "Ts", "P", "Q", "O"),
                   SC=c("nu1", "nu2", "m12", "m21", "Ts", "Tsc", "O"),
                   SC2N=c("nu1", "nu2", "hrf", "m12", "m21", "Ts", "Tsc", "Q", "O"),
                   SCG=c("nu1", "nu2", "b1", "b2", "m12", "m21", "Ts", "Tsc", "O"),
                   SC2N2m=c("nu1", "nu2", "hrf", "m12", "m21", "me12", "me21", "Ts", "Tsc", "P", "Q", "O" ),
                   SC2NG=c("nu1", "nu2", "b1", "b2", "hrf", "m12", "m21", "Ts", "Tsc", "Q", "O"),
                   SC2m=c("nu1", "nu2", "m12", "m21", "me12", "me21", "Ts", "Tsc", "P", "O"),
                   SC2mG=c("nu1", "nu2", "b1", "b2", "m12", "m21", "me12", "me21", "Ts", "Tsc", "P", "O"),
                   SC2N2mG=c("nu1", "nu2", "b1", "b2", "hrf", "m12", "m21", "me12", "me21", "Ts", "Tsc", "P", "Q", "O")
                   
)
```

```{r writePythonScript, eval=FALSE}
write.table("#!/bin/bash","../scripts/make_best_plots_dadi.sh",
            quote=FALSE,row.names = FALSE,col.names = FALSE)
write.table("cd \"${0%/*}\"","../scripts/make_best_plots_dadi.sh",
            append=TRUE, quote=FALSE,row.names = FALSE,col.names = FALSE)
write.table("cd ../fwsw_results/dadi_results/","../scripts/make_best_plots_dadi.sh",
            append=TRUE, quote=FALSE,row.names = FALSE,col.names = FALSE)
pyt_tab<-apply(outTable,1,function(dat,params_order){
  pord<-params_order[[dat["Model"]]]
  pop1<-gsub("(\\w+)_(\\w+)","\\1",dat["PopPair"])
  if(pop1=="FLFW") pop1<-"FLLG"
  pop2<-gsub("(\\w+)_(\\w+)","\\2",dat["PopPair"])
  parms<-gsub(" ","",dat[pord])
  write.table(rbind(c("python ../../scripts/253_plot_best_dadi.py",
                pop1,pop2,dat["Model"],parms)),"../scripts/make_best_plots_dadi.sh",
              append=TRUE,sep=" ",quote=FALSE,row.names = FALSE,col.names = FALSE)
},params_order=params_order)
```



## References



