---
title: 'Supplemental Material #1: Supplemental Population Structure Analyses for 
"Multiple colonizations of freshwater by the Gulf pipefish reveal a shared genomic signature of adaptation"'
preprint: false
author: 
  - name: Sarah P. Flanagan
    affilnum: 1
    corresponding: true
    email: spflanagan.phd@gmail.com
  - name: Emily Rose
    affilnum: 2
  - name: Adam Jones
    affilnum: 3
affiliation:
  - affilnum: 1
    affil: School of Biological Sciences, University of Canterbury, 4800 Private Bag, Christchurch 8140 New Zealand
  - affilnum: 2
    affil: Department of Biology, The University of Tampa, Tampa, FL 33606 USA
  - affilnum: 3
    affil: Department of Biological Sciences, University of Idaho, Moscow, ID 83844 USA
abstract: >
  This document includes supplementary material for the paper. 
  In this document, we show additional methods and results that were involved in the analysis of population structure analyses (Fsts, PCAdapt, Admixture, and Treemix).
  These analyses ultimately contributed to Figure 1 in the main text.
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
bibliography: 
  - programs.bib
  - RADseq.bib
output:
  bookdown::pdf_document2:
    toc: true
    toc_depth: 2
    fig_caption: yes
    keep_tex: yes
    number_sections: no
    template: manuscript.latex
  html_document: null
  word_document: null
fontsize: 11pt
capsize: normalsize
csl: molecular-ecology.csl
documentclass: article
spacing: singlespacing
editor_options: 
  chunk_output_type: console
---



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,out.extra='',fig.pos="H",
                      warning = FALSE,message = FALSE,
                      dev='png',dpi=300)
knitr::opts_knit$set(root.dir='../fwsw_results/')
```
```{r source}
source("../../gwscaR/R/gwscaR.R")
source("../../gwscaR/R/gwscaR_plot.R")
source("../../gwscaR/R/gwscaR_utility.R")
source("../../gwscaR/R/gwscaR_fsts.R")
source("../../gwscaR/R/gwscaR_popgen.R")
source("../../gwscaR/R/vcf2dadi.R")
source("../R/203_treemix_plotting_funcs.R")#I've modified these functions
library(knitr)
library(scales)
library(kableExtra)
library(vegan)
```

```{r popSetup}
pop.list<-c("TXSP","TXCC","TXFW","TXCB","LAFW","ALST","ALFW","FLSG","FLKB",
	"FLFD","FLSI","FLAB","FLPB","FLHB","FLCC","FLLG")
pop.labs<-c("TXSP","TXCC","TXFW","TXCB","LAFW","ALST","ALFW","FLSG","FLKB",
            "FLFD","FLSI","FLAB","FLPB","FLHB","FLCC","FLFW")
fw.list<-c("TXFW","LAFW","ALFW","FLLG")
sw.list<-c("TXSP","TXCC","TXCB","ALST","FLSG","FLKB",
	"FLFD","FLSI","FLAB","FLPB","FLHB","FLCC")
lgs<-c("LG1","LG2","LG3","LG4","LG5","LG6","LG7","LG8","LG9","LG10","LG11",
	"LG12","LG13","LG14","LG15","LG16","LG17","LG18","LG19","LG20","LG21",
	"LG22")
lgn<-seq(1,22)
all.colors<-c(rep("black",2),"#2166ac","black","#2166ac","black","#2166ac",
        rep("black",8),"#2166ac")
#grp.colors<-c('#e41a1c','#377eb8','#4daf4a','#984ea3','#ffff33','#f781bf')
grp.colors<-c('#762a83','#af8dc3','#e7d4e8','#d9f0d3','#7fbf7b','#1b7837')
col_vector<-c(red='#e6194b', green='#3cb44b', blue='#4363d8',yellow='#ffe119', 
              cyan='#46f0f0',orange='#f58231', teal='#008080', purple='#911eb4',  
              magenta='#f032e6', lime='#bcf60c', pink='#fabebe',  lavendar='#e6beff', 
              brown='#9a6324', olive='#808000', apricot='#ffd8b1',maroon='#800000', 
              mint='#aaffc3', navy='#000075', beige='#fffac8', grey='#808080', 
              white='#ffffff', black='#000000')

col_vector<-c('#762a83','#762a83',"#2166ac",'#762a83',"#2166ac",'#af8dc3',
              "#2166ac",'#e7d4e8','#e7d4e8','#e7d4e8','#e7d4e8','#7fbf7b',
              '#1b7837','#1b7837','#1b7837',"#2166ac")

ppi<-data.frame(Pop=pop.labs,cols = col_vector,
                pch=rep(c(15,16,17,18),4))
ppi$pch[grep("FW",ppi$Pop)]<-c(15,16,17,18)
```

\newpage

# Sequencing library preparation additional methods {-}

DNA was extracted from the frozen tissues of the pipefish using an ethanol-precipitation method with an RNase-A step (Qiagen PureGene). The ensuing DNA samples were evaluated for quality using a Nanodrop spectrophotometer, a Qubit 2.0 Fluorometer (Life Technologies), and by examination of the extracted DNA on a 1% agarose gel.

We then followed published double-digest restriction site associated DNA sequencing protocols. Specifically, we used the single indexing scheme from the single-digest RAD-seq protocols [@bairdRapidSNPDiscovery2008; @millerRapidCosteffectivePolymorphism2007], and the enzymes PstI-HF (New England Biolabs) and MboI (New England Biolabs). We digested 1 $\mu$g of each DNA sample with the two enzymes, and 250ng of each digest was used in the ligation step, when barcodes were added to individual samples. Following the ligation, 96 samples were pooled to form a single library, and we selected for 250-500bp fragments using a 1% agarose gel. Fragments containing the sequencing primers were selected and amplified in four separate PCRs per sequencing library, each with 12 cycles, and those PCRs were pooled. All purification steps were done with AMPure XP beads (Agilent), except for the gel purification following the size selection, in which we used the Zymo Gel purification kit.  For a detailed description of these methods, please see [@flanaganPopulationGenomicsReveals2016]. 

As described in the main manuscript, we analyzed the resulting RAD loci to create two sets of single nucleotide polymorphisms (SNPs):

1. A dataset with all 16 populations, generated from all pairwise comparisons of populations, containing SNPs found in every population, in 75% of individuals, and with a minor allele frequency of at least 5%.

2. A dataset containing only the 4 freshwater populations (TXFW, LAFW, ALFW, FLFW) and their nearest saltwater populations (TXCC, ALST, FLCC -- note ALST is the nearest neighbor to both ALFW and LAFW). This dataset also contains SNPs found in 75% of individuals with a minor allele frequency of at least 5%. 

This supplement focuses on providing supplementary methods and results of the analysis of dataset #1, which was used to investigate population structure. Supplements 2 and 3 focus on the second dataset.


# Analysis of population structure {-}

## Phenotypic variation


```{r organizeData, eval=FALSE}
raw.pheno<-read.table("../sw_results/popgen.pheno.txt", sep="\t", header=T)
	raw.pheno$PopID<-gsub("(\\w{4})\\w+","\\1",raw.pheno$ID)
	raw.pheno<-raw.pheno[raw.pheno$PopID %in% pop.list,]
	raw.pheno$sex<-gsub("\\w{4}(\\w)\\w+","\\1",raw.pheno$ID)
	raw.pheno$TailLength<-raw.pheno$std.length-raw.pheno$SVL
	raw.pheno$HeadLength<-raw.pheno$HeadLength-raw.pheno$SnoutLength

fem.pheno<-raw.pheno[raw.pheno$sex %in% c("F","D"),-8]
	fem.pheno<-fem.pheno[,c(11,1,10,2,12,4,5,6,7,8,9)]
	fem.pheno<-fem.pheno[order(match(fem.pheno$PopID,pop.list)),]
	write.table(fem.pheno,"fem.pheno.txt",sep='\t',row.names=F,col.names=T,
		quote=F)
	
mal.pheno<-raw.pheno[raw.pheno$sex %in% c("P","N"),-8]
	mal.pheno<-mal.pheno[,c(11,1,10,2,12,4,5,6,7)]
	mal.pheno<-mal.pheno[order(match(mal.pheno$PopID,pop.list)),]
	write.table(mal.pheno,"mal.pheno.txt",sep='\t',row.names=F,col.names=T,
		quote=F)
```
```{r readMalFem}
fem.pheno<-read.table("fem.pheno.txt",header=T)
	fem.pheno<-fem.pheno[!is.na(fem.pheno$BandNum),]
mal.pheno<-read.table("mal.pheno.txt",header=T)
```

We photographed every fish collected during this study, and analyzed the photographs of the `r nrow(fem.pheno)` females and `r nrow(mal.pheno)` males. We measured body traits (tail length, snout-vent length, body depth, snout length, snout depth, and head length without snout length) in all individuals. We also measured band area and band number in feamles. To investigate phenotypic variation among populations, we used vegan version `r packageVersion("vegan")` [@oksanenVeganCommunityEcology2015] to conduct principal components analyses (using the function `rda`). We analyzed males and females separately. We also conducted a principal components analysis on the two female band traits alone. 

```{r PCA}
fem.pheno$PopID<-factor(fem.pheno$PopID)
fem.pheno<-fem.pheno[!is.na(fem.pheno$BandNum),]
mal.pheno$PopID<-factor(mal.pheno$PopID)
bands.pcdat<-fem.pheno[!is.na(fem.pheno$BandNum),
	c("PopID","ID","MBandArea","BandNum")]
# run pcas
band.pca<-rda(bands.pcdat[,3:4])
fem.pheno.pca<-rda(fem.pheno[,4:9])
mal.pheno.pca<-rda(mal.pheno[,4:9])

```
```{r extractEigenvalue}
####extract eigenvalue
band.eig<-band.pca$CA$eig
band.pc<-band.eig/sum(band.eig)*100

#extract PC scores
band.u<-data.frame(bands.pcdat[,1:2],
                   "BandPC1"=band.pca$CA$u[,1],stringsAsFactors=F)
band.u.sep<-split(band.u, band.u[,1])
band.u.new<-rbind(band.u.sep$TXSP,band.u.sep$TXCC,band.u.sep$TXCB,
	band.u.sep$ALST,band.u.sep$FLSG,band.u.sep$FLKB,
	band.u.sep$FLFD,band.u.sep$FLSI,band.u.sep$FLAB,
	band.u.sep$FLPB,band.u.sep$FLHB,band.u.sep$FLCC)

fem.pheno.eig<-fem.pheno.pca$CA$eig
fem.pheno.pc<-fem.pheno.eig/sum(fem.pheno.eig)*100

#extract PC scores
fem.pheno.u<-data.frame(fem.pheno[,1:2],
	"FemBodyPC1"=fem.pheno.pca$CA$u[,1],stringsAsFactors=F)
fem.u.sep<-split(fem.pheno.u, fem.pheno.u[,1])
fem.u.new<-rbind(fem.u.sep$TXSP,fem.u.sep$TXCC,fem.u.sep$TXCB,
	fem.u.sep$ALST,fem.u.sep$FLSG,fem.u.sep$FLKB,
	fem.u.sep$FLFD,fem.u.sep$FLSI,fem.u.sep$FLAB,
	fem.u.sep$FLPB,fem.u.sep$FLHB,fem.u.sep$FLCC)

mal.pheno.eig<-mal.pheno.pca$CA$eig
mal.pheno.pc<-mal.pheno.eig/sum(mal.pheno.eig)*100

#extract PC scores
mal.u<-data.frame(mal.pheno[,1:2],"MalBodyPC1"=mal.pheno.pca$CA$u[,1],
	stringsAsFactors=F)
mal.u.sep<-split(mal.u, mal.u[,1])
mal.u.new<-rbind(mal.u.sep$TXSP,mal.u.sep$TXCC,mal.u.sep$TXCB,
	mal.u.sep$ALST,mal.u.sep$FLSG,mal.u.sep$FLKB,
	mal.u.sep$FLFD,mal.u.sep$FLSI,mal.u.sep$FLAB,
	mal.u.sep$FLPB,mal.u.sep$FLHB,mal.u.sep$FLCC)

```

```{r PCAplotSetup}
# females
fem.pop<-as.character(bands.pcdat$PopID)
fem.pop[fem.pop=="FLLG"]<-"FLFW"
fem.colors<-as.character(fem.pop)
fem.pch<-as.character(fem.pop)
fw.fem.col<-as.character(fem.pop[fem.pop %in% fw.list])
for(i in 1:length(fem.pop)){
  fem.colors[i]<-as.character(ppi[ppi$Pop %in% fem.pop[i],"cols"])
  fem.pch[i]<-as.numeric(as.character(ppi[ppi$Pop %in% fem.pop[i],"pch"]))
}
fem.pch<-as.numeric(fem.pch)

# males
mal.pop<-as.character(mal.pheno$PopID)
mal.pop[mal.pop=="FLLG"]<-"FLFW"
mal.colors<-as.character(mal.pop)
mal.pch<-as.character(mal.pop)
fw.mal.col<-as.character(mal.pop[mal.pop %in% fw.list])
for(i in 1:length(mal.pop)){
  mal.colors[i]<-as.character(ppi[ppi$Pop %in% mal.pop[i],"cols"])
  mal.pch[i]<-as.character(ppi[ppi$Pop %in% mal.pop[i],"pch"])
}
mal.pch<-as.numeric(mal.pch)


fw.fem.rows<-which(fem.pheno$PopID %in% fw.list)
fw.mal.rows<-which(mal.pheno$PopID %in% fw.list)

```

```{r pcaLoadings}
PCloadings<-data.frame(rbind(mal.pheno.pca$CA$v,
                             fem.pheno.pca$CA$v,
                             cbind(band.pca$CA$v,NA,NA,NA,NA)),
                       stringsAsFactors = FALSE)
PCloadings<-apply(PCloadings,2,round,digits=3)
kable(PCloadings,"latex",booktabs=TRUE,
      caption="Principal components analysis loadings for traits in males and females") %>%
  pack_rows("Males",1,6) %>%
  pack_rows("Females",7,12) %>%
  pack_rows("Bands",13,14) %>%
  kable_styling(latex_options = "HOLD_position")
```


```{r plotPCA,fig.height=8,fig.width=10,fig.keep='last',fig.path="../figs/",fig.cap="Principal components analysis of morphological traits in S. scovelli reveals that phenotypic variation among populations is not based on habitat type. The top set of panels show the results of the PCA with all 16 populations, color-coded by populations and point shape. The bottom set of panels show the same PCA results, but with different x- and y-axis scaling and without the saltwater populations plotted, to facilitate visualizing the differences among saltwater populations. The left panels show male body traits (SVL, tail length, trunk depth, head length, snout length, and snout depth), the middle panels show those same traits in females, and the right panels show the female band traits (band number and band area)."}
ptCex<-2

par(mfrow=c(2,3),oma=c(2,2,2,2),mar=c(2,2,2,2),lwd=1.3)
mp<-plot(mal.pheno.pca,type="n",xlim=c(-3,3),ylim=c(-8.2,4)
	,xlab="",ylab="",las=1,cex.axis=1.5)
points(mal.pheno.pca,col=alpha(mal.colors,0.5),cex=ptCex,pch=mal.pch)
mtext(paste0("PC1 (",round(mal.pheno.pc[1],2),"%)"),1,line=2)
mtext(paste0("PC2 (",round(mal.pheno.pc[2],2),"%)"),2,line=2.5)
legend("top",bty='n',c("Male Body Traits"),cex=1.5)

fp<-plot(fem.pheno.pca,type="n",xlab="",ylab="",las=1,cex.axis=1.5,ylim=c(-4,12),
	xlim=c(-3,3))
points(fem.pheno.pca,col=alpha(fem.colors,0.5),cex=ptCex,pch=fem.pch)
mtext(paste0("PC1 (",round(fem.pheno.pc[1],2),"%)"),1,line=2)
mtext(paste0("PC2 (",round(fem.pheno.pc[2],2),"%)"),2,line=2.5)
legend("top",bty='n',c("Female Body Traits"),cex=1.5)

bp<-plot(band.pca,type="n",xlab="",ylab="",las=1,
         cex.axis=1.5,xlim=c(-2,2),ylim=c(-3,1))
points(band.pca,pch=fem.pch,col=alpha(fem.colors,0.5),cex=ptCex)
mtext(paste0("PC1 (",round(band.pc[1],2),"%)"),1,line=2)
mtext(paste0("PC2 (",round(band.pc[2],2),"%)"),2,line=2.5)
legend("top",bty='n',c("Female Band Traits"),cex=1.5)


plot(mp$sites[fw.mal.rows,],type="n",xlab="",ylab="",las=1,cex.axis=1.5)
abline(h=0,lty=3)
abline(v=0,lty=3)
points(mp$sites[fw.mal.rows,],xlim=c(-0.1,0.1),ylim=c(-.2,.2),
	col=alpha(mal.colors[fw.mal.rows],0.5),cex=ptCex,pch=mal.pch[fw.mal.rows])
mtext(paste0("PC1 (",round(mal.pheno.pc[1],2),"%)"),1,line=2)
mtext(paste0("PC2 (",round(mal.pheno.pc[2],2),"%)"),2,line=2.5)


plot(fp$sites[fw.fem.rows,],type="n",xlab="",ylab="",las=1,
	cex.axis=1.5,ylim=c(-4,12),xlim=c(-3,3))
abline(h=0,lty=3)
abline(v=0,lty=3)
points(fp$sites[fw.fem.rows,],
	col=alpha(fem.colors[fw.fem.rows],0.5),cex=ptCex,
	pch=fem.pch[fw.fem.rows])
mtext(paste0("PC1 (",round(fem.pheno.pc[1],2),"%)"),1,line=2)
mtext(paste0("PC2 (",round(fem.pheno.pc[2],2),"%)"),2,line=2.5)


plot(bp$sites[fw.fem.rows,],type="n",xlab="",ylab="",las=1,
	cex.axis=1.5,xlim=c(-2,2),ylim=c(-3,1))
points(bp$sites[fw.fem.rows,],
	pch=fem.pch[fw.fem.rows],col=alpha(fem.colors[fw.fem.rows],0.5),
	cex=ptCex)
mtext(paste0("PC1 (",round(band.pc[1],2),"%)"),1,line=2)
mtext(paste0("PC2 (",round(band.pc[2],2),"%)"),2,line=2.5)
abline(h=0,lty=3)
abline(v=0,lty=3)

par(fig = c(0, 1, 0, 1), oma=c(2,1,0,1), 
    mar = c(0, 0, 0, 0), new = TRUE, cex=1)
plot(0, 0, type = "n", bty = "n", xaxt = "n", yaxt = "n")
legend("top", legend=ppi$Pop, 
	col=as.character(ppi$col),
	pt.cex=ptCex,bty='n',pch=ppi$pch, ncol=8)
```


The principal components analyses revealed very little population structuring in traits (Figure \@ref(fig:plotPCA)). In all three principal components analyses, the first axis explained over 90% of the variation in traits. In both males and females, the body trait with the highest loading on the first principal components axis was Tail Length (males: `r max(mal.pheno.pca$CA$v[,1])`, females: `r max(fem.pheno.pca$CA$v[,1])`), with snout-vent length as a close second (males: `r mal.pheno.pca$CA$v["SVL",1]`, females: `r fem.pheno.pca$CA$v["SVL",1]`). In both males and females, SVL had the highest loading with tail length having the second largest magnitude (but negative; Table \@ref(tab:pcaLoadings)). For the bands, band number had the largest absolute contribution to the first axis, with mean band number contributing most to the second PC axis (Table \@ref(tab:pcaLoadings)). Together, these results suggest that body size was the largest driver of variation among individuals, but the lack of clear population clusters suggests that populations did not have consistent differences in body size or band traits.

## Minor allele frequencies

We calculated the minor allele frequency for all 7933 SNPs in the full dataset for each population. We achieved this using the function `calc.afs.vcf` in the `gwscaR` package, which calculates the frequencies for the reference and alternative allele. We then extracted and plotted the frequency of the least-common allele in each population (Figure \@ref(fig:maf)).

These histograms demonstrate that the two most isolated freshwater populations (FLFW and TXFW) have minor allele frequencies skewed towards zero, in particular compared to many of the saltwater populations in the Gulf of Mexico (e.g., FLSG, TXCB).

```{r calcAFS, eval=FALSE}
locus.info<-colnames(ful_vcf[1:9])
fw.afs<-lapply(fw.list,function(pop,vcf){
  this.vcf<-cbind(vcf[,locus.info],vcf[,grep(pop,colnames(vcf))])
  this.afs<-do.call(rbind,apply(this.vcf,1,calc.afs.vcf))
}, vcf=ful_vcf)
names(fw.afs)<-c("TXFW","LAFW","ALFW","FLFW")
sw.afs<-lapply(sw.list,function(pop,vcf){
  this.vcf<-cbind(vcf[,locus.info],vcf[,grep(pop,colnames(vcf))])
  this.afs<-do.call(rbind,apply(this.vcf,1,calc.afs.vcf))
},vcf=ful_vcf)
names(sw.afs)<-sw.list
all.afs<-c(fw.afs,sw.afs)
minAF<-lapply(all.afs,function(x){
  mins<-apply(x,1,function(row){ return(min(row[4],row[6])) } )
  return(as.numeric(mins))
})
names(minAF)<-names(all.afs)

```

```{r minorAFplot,fig.height=10,fig.width=8,dpi=300,fig.keep='last', fig.path="../figs/",eval=FALSE}

par(mfrow=c(4,4),mar=c(2,2,1,0),oma=c(2,3,0.5,0.5))
for(i in 1:length(pop.labs)){
  if(pop.labs[i] %in% names(fw.afs)){
    color<-"cornflowerblue"
  }else{
    color<-"black"
  }
  hist(minAF[[pop.labs[i]]],ylab="",xlab="",main="",
       xlim=c(0,0.5),ylim=c(0,10000),axes=F,col=color,
       breaks=seq(0,0.5,0.05))
  axis(1,pos=0,cex.axis=2)
  if(i %in% c(1,5,9,13)){
    axis(2,pos=0,las=1,cex.axis=2,labels = seq(0,10,2),at=seq(0,10000,2000))
  }else{
    axis(2,pos=0,las=1,labels = FALSE,cex.axis=2)
    }
  mtext(pop.labs[i],3,col=color,cex=2*0.75,line=-1)
}
mtext("Minor Allele Frequency",1,outer=TRUE,cex=1.75*0.75)
mtext("Number of SNPs (x 1000)",2,outer = TRUE,cex=1.75*0.75,line=1)

```

```{r maf, out.width='80%',fig.cap="Minor allele frequency distributions of the full dataset (7433 SNPs) for each population. Freshwater populations are plotted in blue. The histograms show the number of SNPs with various frequencies of the reference alleles. All populations are skewed towards having small minor allele frequencies, but the TXFW and FLFW have additional reductions in genetic variation."}
knitr::include_graphics('../figs/minorAFplot-1.png')
```


## Fst Values

 

```{r readStacksFstSumm}
full_fsts<-read.delim("stacks/populations_whitelist/batch_2.fst_summary.tsv",
                      row.names = 1)
full_fsts<-rbind(full_fsts,TXSP=rep(NA,ncol(full_fsts))) #add the final row
Tfull_fsts<-t(full_fsts)
full_fsts[lower.tri(full_fsts)]<-Tfull_fsts[lower.tri(Tfull_fsts)] # now it's symmetric
full_fsts<-full_fsts[pop.list,pop.list]
colnames(full_fsts)<-rownames(full_fsts)<-pop.labs
fst_mat<-as.matrix(full_fsts)
```

To visualize population structuring based on genotypes, we investigated pairwise $F_{ST}$ values output by Stacks [@catchenStacksAnalysisTool2013]. The pairwise $F_{ST}$ values ranged from `r min(fst_mat,na.rm=TRUE)` to `r max(fst_mat,na.rm=TRUE)`, with a mean of `r mean(fst_mat,na.rm=TRUE)`. The patterns in the $F_{ST}$ values were also compared to the covariances calculated by Treemix ([@pickrellInferencePopulationSplits2012]), using the unrooted population tree (see \@ref(treemix) for more detail). We found that the most similar populations (i.e., with low $F_{ST}$ and high covariance) were located nearest to each other geographically (Figure \@ref(fig:fsts)), consistent with the finding of an isoaltion-by-distance pattern in [@flanaganPopulationGenomicsReveals2016]. The freshwater populations show somewhat elevated $F_{ST}$ values and decreased covariances, especially the freshwater populations in Florida and Texas.

```{r fst_heatmaps,fig.width=8,fig.height=5,eval=FALSE}
library(lattice); library(grid); library(RColorBrewer)
poporder<-read.delim("treemix/poporder")
colors<-poporder$colors
poporder<-poporder$poporder

nr<-treemix.cov.plot("treemix/fwsw_k100b",poporder)
dimnames(nr)[[1]]<-dimnames(nr)[[2]]<-pop.labs
# colors
colors<-c("black","darkgrey","grey","lightgrey","cornflowerblue")
pal<-colorRampPalette(colors)
ncol=80
cols<-pal(ncol)
rev.colors<-c("cornflowerblue","lightgrey","grey","darkgrey","black")
rev.pal<-colorRampPalette(rev.colors)
rev.cols<-rev.pal(ncol)

hm.height<-list(x=2,units="in")#2.2/3.8
hm.width<-list(x=2.4,units="in")#2.4 in RStudio/3.9

heatmaps.name<-"../figs/fst_heatmaps.png"

png(heatmaps.name,height=5,width=8,units="in",res=300)

fst.lv<-levelplot(as.matrix(fst_mat),col.regions=cols,alpha.regions=0.7,
                  scales = list(x=list(rot=90),tck = 0),xlab="",ylab="")
print(fst.lv,split=c(1,1,2,1),more=TRUE,panel.width=hm.width,
      panel.height=hm.height,cex=2)
trellis.focus("legend", side="right", clipp.off=TRUE, highlight=FALSE)
grid.text(expression(italic(F)[ST]), 0.2, 0, hjust=0.5, vjust=1.2,gp=gpar(cex=0.75))
trellis.unfocus()

nr.lv<-levelplot(nr,col.regions=cols,alpha.regions=0.7,
                 scales = list(x=list(rot=90),tck = 0),xlab="",ylab="")
print(nr.lv,split=c(2,1,2,1),more=FALSE,newpage=FALSE,panel.width=hm.width,
      panel.height=hm.height,cex=2)
trellis.focus("legend", side="right", clipp.off=TRUE, highlight=FALSE)
grid.text("Treemix", 0.2, 0, hjust=0.5, vjust=1.2,gp=gpar(cex=0.75))
trellis.unfocus()

dev.off()

```

```{r fsts, out.width='95%',fig.cap="Heatmaps depicting population structure. In all graphs, dark colors depict similarity between populations and light grey and blue depict populations with high differentiation. The left panel shows pairwise $F_{ST}$ values calculated by the populations module in Stacks (Catchen et al. 2013). The right panel shows covariances between populations as calculated by TreeMix (Pickrell and Pritchard 2012)."}
knitr::include_graphics('../figs/fst_heatmaps.png')
```



## PCAdapt {-}

Principal components analysis was implemented using PCAdapt [@luuPcadaptPackagePerform2017]. We initially ran the analysis on the full dataset of 7433 SNPs, visualizing 20 principal components axes to determine the optimal number of axes to retain that explain the majority of the variation among samples and populations. By inspecting the plot of the proportion of variation explained by each axis (Figure \@ref(fig:pcadaptScree)), we determined that retaining 7 axes was appropriate for our dataset.

```{r loadPCAdapt}
library(pcadapt)
```

```{r pcadaptChoose,results='hide',message=FALSE}
filename<-read.pcadapt("filter_rad_20191014@1654/14_filtered/radiator_data_20191014@1710.vcf",
                       type="vcf")
x<-pcadapt(filename, K=20)
```
```{r pcadaptScree,fig.cap="Scree plot from PCAdapt with 20 principal components axes retained."}
plot(x,option="screeplot") #K=7
```

```{r pcadaptAnalyze}
pa<-pcadapt(filename,K=7)
saveRDS(pa,"fwsw_all_pcadapt.RDS")
pa.props<-round((pa$singular.values/sum(pa$singular.values))*100,2)
kable(t(pa.props),row.names = FALSE,col.names = 1:7,booktabs=TRUE,
      caption="Proportion of variation explained by all 7 of the retained PC axes in PCAdapt") %>%
  kable_styling(latex_options="hold_position")
```
```{r pcadaptSummarize}
ind_dat<-read.table(
  "filter_rad_20191014@1654/14_filtered/individuals.qc.stats_20191014@1654.tsv",
  header=T, stringsAsFactors = F)
pops<-ind_dat$STRATA	
grp<-pops
grp[grp=="TXFW" | grp=="LAFW" | grp=="ALFW" | grp=="FLLG"]<-"freshwater"
grp[grp!="freshwater"]<-"saltwater"

#colors
pap<-data.frame(Pop=pops,cols=pops,pch=pops,grp=grp,stringsAsFactors = F)
pap$Pop[pap$Pop == "FLLG"]<-"FLFW"
for(i in 1:nrow(pap)){
  pap[i,"cols"]<-as.character(ppi[ppi$Pop %in% pap[i,"Pop"],"cols"])
}
for(i in 1:nrow(pap)){
  pap[i,"pch"]<-as.numeric(ppi[ppi$Pop %in% pap[i,"Pop"],"pch"])
}
write.table(pap,"pcadapt_colp.txt",col.names=TRUE,sep='\t',quote=F)

```

These seven axes explained `r sum(pa.props[1:7])`% of the variation in the dataset (Table \@ref(tab:pcadaptAnalyze)), and the first six axes accounted for `r sum(pa.props[1:6])`% of the variation. We inspected the location of individual samples in the principal components space and observed clear population structuring (Figure \@ref(fig:PCAdaptPlot)) in line with the isolation-by-distance patterns in [@flanaganPopulationGenomicsReveals2016] and in the ADMIXTURE (\@ref(admixture)) and Treemix analyses (\@ref(treemix)). The freshwater populations showed clear divergence between the FLFW population and the Texas, Alabama, and Louisiana populations (Figure \@ref(fig:PCAdaptPlot), bottom left panel), and the Texas freshwater population was distinct from the Alabama and Louisiana populations (Figure \@ref(fig:PCAdaptPlot), bottom center panel). The Alabama and Louisiana freshwater populations were unable to be distinguished from each other in any of the first six principal components axes (Figure \@ref(fig:PCAdaptPlot)).

```{r plot6pcsPCAdapt,fig.height=8,fig.width=10.5,fig.pos='H',fig.path="../figs/", fig.show=FALSE, fig.cap="Principal components analysis of genotypes in S. scovelli.",eval=FALSE}

par(mfrow=c(2,3),oma=c(2,2,2,2),mar=c(2,2,2,2))
plot(pa$scores[,1],pa$scores[,2],col=alpha(pap$cols,0.5),bg=alpha(pap$cols,0.75),
     pch=as.numeric(pap$pch),	cex=1.5)
mtext(paste("PC1 (",pa.props[1],"%)",sep=""),1,line = 2,cex=0.75)
mtext(paste("PC2 (",pa.props[2],"%)",sep=""),2,line = 2,cex=0.75)
plot(pa$scores[,3],pa$scores[,4],col=alpha(pap$cols,0.5),
     bg=alpha(pap$cols,0.75),pch=as.numeric(pap$pch),cex=1.5)
mtext(paste("PC3 (",pa.props[3],"%)",sep=""),1,line = 2,cex=0.75)
mtext(paste("PC4 (",pa.props[4],"%)",sep=""),2,line = 2,cex=0.75)
plot(pa$scores[,5],pa$scores[,6],col=alpha(pap$cols,0.5),
     bg=alpha(pap$cols,0.75),pch=as.numeric(pap$pch),cex=1.5)
mtext(paste("PC5 (",pa.props[5],"%)",sep=""),1,line = 2,cex=0.75)
mtext(paste("PC6 (",pa.props[6],"%)",sep=""),2,line = 2,cex=0.75)
plot(pa$scores[grp=="freshwater",1],pa$scores[grp=="freshwater",2],
     col=alpha(pap$cols[pap$grp=="freshwater"],0.5),
     bg=alpha(pap$cols[pap$grp=="freshwater"],0.75),
     pch=as.numeric(pap$pch[pap$grp=="freshwater"]),
	cex=1.5)
mtext(paste("PC1 (",pa.props[1],"%)",sep=""),1,line = 2,cex=0.75)
mtext(paste("PC2 (",pa.props[2],"%)",sep=""),2,line = 2,cex=0.75)
plot(pa$scores[grp=="freshwater",3],pa$scores[grp=="freshwater",4],
     col=alpha(pap$cols[pap$grp=="freshwater"],0.5),
     bg=alpha(pap$cols[pap$grp=="freshwater"],0.75),
     pch=as.numeric(pap$pch[pap$grp=="freshwater"]),
	cex=1.5)
mtext(paste("PC3 (",pa.props[3],"%)",sep=""),1,line = 2,cex=0.75)
mtext(paste("PC4 (",pa.props[4],"%)",sep=""),2,line = 2,cex=0.75)
plot(pa$scores[grp=="freshwater",5],pa$scores[grp=="freshwater",6],
     col=alpha(pap$cols[pap$grp=="freshwater"],0.5),
     bg=alpha(pap$cols[pap$grp=="freshwater"],0.75),
     pch=as.numeric(pap$pch[pap$grp=="freshwater"]),
	cex=1.5)
mtext(paste("PC5 (",pa.props[5],"%)",sep=""),1,line = 2,cex=0.75)
mtext(paste("PC6 (",pa.props[6],"%)",sep=""),2,line = 2,cex=0.75)

par(fig = c(0, 1, 0, 1), oma=c(2,1,0,1), mar = c(0, 0, 0, 0), new = TRUE,
	cex=1)
plot(0, 0, type = "n", bty = "n", xaxt = "n", yaxt = "n")

legend("top", legend=ppi$Pop, pch=as.numeric(ppi$pch), pt.cex=1.5,cex=0.85,
       col=alpha(ppi$cols, 0.5),pt.bg=alpha(ppi$cols,0.25), ncol=8,bty='n')
```

```{r PCAdaptPlot, out.width='90%',fig.cap="Principal components analysis of genotypes in the Gulf pipefish reveals population structure due to geographic distance and habitat type. The top set of panels show the results of the PCA with all 16 populations, color-coded by populations and point shape. The bottom set of panels show the same PCA results, without the saltwater populations plotted (and zoomed in), to facilitate visualizing the differences among saltwater populations."}
knitr::include_graphics('../figs/plot6pcsPCAdapt-1.png')
```


## Admixture {#admixture}

We used ADMIXTURE to estimate the ancestry of individuals assuming $K=1$ through $K=16$ independent clusters. We determined the optimal number of clusters to include by choosing the K associated with the lowest cross-validation error, which was either $K=5$ or $K=7$ (Figure \@ref(fig:admixScree)).

```{r admixScree,fig.cap="Admixture screeplot for K=1 through K=16. The cross-validation error (CV) is shown on the y-axis."}
admixK<-read.delim("admixture/K_CVs.txt",header = FALSE)
admixK$K<-as.numeric(gsub(".*\\(K=(\\d+)\\).*","\\1",admixK$V1))
admixK$CV<-as.numeric(gsub("^.*\\: (\\d+\\.\\d+)$","\\1",admixK$V1))

admixK<-admixK[order(admixK$K),]

plot(admixK$K,admixK$CV,pch=19,type = "b",lty=1,xlab = "K",ylab="CV",las=1,lwd=2)

```

To visualize the effect of our choice of $K$ on the assignment of individuals to clusters, we visualized the admixture plots for $K=2$ through $K=7$ (Figures \@ref(fig:admixK2),\@ref(fig:admixK3),\@ref(fig:admixK4),\@ref(fig:admixk5),\@ref(fig:admixk6),\@ref(fig:admixk7)). 

```{r admixSetup}
library(RColorBrewer)
famfile<-"admixture/fwsw_all_filt.fam"

poporderFile<-"treemix/poplist"
poporderDF<-read.table(poporderFile,col.names = c("Pop"),stringsAsFactors = F)
poporderDF$orderNum<-1:nrow(poporderDF)
```

```{r plotAdmixFxn}
admixPlotting<-function(qfile,K,famfile="admixture/fwsw_all_filt.fam",
                        poporder=poporderDF){
  # read files in 
  famTable<- read.table(famfile, col.names = 
                        c("Pop","Ind","Father","Mother","Sex","phenotype"),
                      stringsAsFactors = F)[1:2]

  qtbl<-read.table(qfile,stringsAsFactors = F)
  
  # create useful tables
  mergedAdmixtureTable <- cbind(qtbl, famTable)
  mergedAdmixTabOrderNs <- merge(mergedAdmixtureTable,poporder,by="Pop")
  ordered <- mergedAdmixTabOrderNs[order(mergedAdmixTabOrderNs$orderNum),]
  
  plotting.structure(ordered[,1:(ncol(ordered)-2)],k = K,
                     pop.order = poporder$Pop,make.file = FALSE)
  admix<-ordered[,1:(ncol(ordered)-2)]
  return(admix)
}
```

```{r admixK2,fig.height=1,fig.width=8,fig.cap="Admixture plot for K=2. The colors represent different genetic populations."}
admixK2<-admixPlotting("admixture/fwsw_all_filt.2.Q",2)
```

```{r admixK3,fig.height=1,fig.width=8,fig.cap="Admixture plot for K=3. The colors represent different genetic populations."}
admixK3<-admixPlotting("admixture/fwsw_all_filt.3.Q",3)
```
```{r admixK4,fig.height=1,fig.width=8,fig.cap="Admixture plot for K=4. The colors represent different genetic populations."}
admixK4<-admixPlotting("admixture/fwsw_all_filt.4.Q",4)
```

```{r admixk5,fig.height=1,fig.width=8,fig.cap="Admixture plot for K=5. The colors represent different genetic populations."}
admixK5<-admixPlotting("admixture/fwsw_all_filt.5.Q",5)
write.table(admixK5,"admixture/admixK5.txt",sep = '\t',
            quote = FALSE,col.names = TRUE,row.names = FALSE)
```

```{r admixk6,fig.height=1,fig.width=8,fig.cap="Admixture plot for K=6. The colors represent different genetic populations."}
admixK6<-admixPlotting("admixture/fwsw_all_filt.6.Q",6)
```

```{r admixk7,fig.height=1,fig.width=8,fig.cap="Admixture plot for K=7. The colors represent different genetic populations."}
admixK7<-admixPlotting("admixture/fwsw_all_filt.7.Q",7)
write.table(admixK7,"admixture/admixK7.txt",sep = '\t',quote = FALSE,col.names = TRUE,row.names = FALSE)
```


# Analysis of demographic structure {-}


## Treemix analysis {#treemix}

We used Treemix [@pickrellInferencePopulationSplits2012] to infer population splits and migration events among all 16 populations. Throughout the analyses we used blocks of 100 SNPs. 

```{r treemixPoporder}
poporder<-c("TXSP","TXCC","TXFW","TXCB","LAFW","ALST",
            "ALFW","FLSG","FLKB","FLFD","FLSI","FLAB",
            "FLPB","FLHB","FLCC","FLLG")
colors<-poporder
colors[colors %in% "FLLG"]<-grp.colors[6]
colors[colors %in% c("FLPB","FLHB","FLCC")]<-grp.colors[6]
colors[colors %in% c("FLAB")]<-grp.colors[5]
colors[colors %in% c("FLSI","FLFD","FLKB","FLSG")]<-grp.colors[3]
colors[colors %in% c("ALST","ALFW","LAFW")]<-grp.colors[2]
colors[colors %in% c("TXSP","TXCC","TXFW","TXCB")]<-grp.colors[1]
write.table(cbind(poporder,colors),"poporder",quote=F,sep='\t')
```

We began by running 100 bootstrap replicates without a root and without migration edges. We then built a consensus tree using PHYLIP v. 3.697 [@felsensteinPHYLIPPhylogenyInference2005] and used that tree (Figure \@ref(fig:treemixConsensus)) to choose the best population to set as a root. We then ran another 100 bootstrap replicates with the chosen population (which was FLAB) as root but with no migration edges. Again, we built a consensus tree with PHYLIP v. 3.697 [@felsensteinPHYLIPPhylogenyInference2005], which we used in the analyses with migration edges. 

```{r treemixConsensus, results='hide', fig.show='true',fig.cap="The consensus tree from running Treemix any migration edges and without a root."}
source("../R/203_treemix_plotting_funcs.R") #I've modified the functions from treemix
library(lattice); library(grid); library(RColorBrewer);library(ape)
poporder<-read.delim("treemix/poporder")
colors<-poporder$colors
poporder<-poporder$poporder
ur<-ape::read.tree("treemix/unrooted_consensus.newick")
rt<-ape::read.tree("treemix/rooted_consensus.newick")
par(mfrow=c(1,2),oma=c(2,2,2,2),mar=c(2,2,2,2))
plot(ur)
plot(rt)
```



### Choosing the optimal number of migration edges

We ran treemix, specifying the consensus tree with FLAB as root as the tree to use, with zero to five migration edges. We can inspect these trees and their residuals to look for patterns.

```{r treemixMigrations, message=FALSE,results='hide',fig.height=10,fig.width=5,fig.cap="Residuals (left) and trees of the treemix runs with migration edges. All models used the consensus tree with FLAB as root."}

par(mfrow=c(6,2),mar=c(2,2,2,2),oma=c(2,2,2,2))
r0<-plot_resid("treemix/migrations/fwsw_FLAB_m0","treemix/poplist",cex = 0.75)
t0<-plot_tree("treemix/migrations/fwsw_FLAB_m0",plotmig = F,scale=T,mbar=F,cex = 0.75)

r1<-plot_resid("treemix/migrations/fwsw_FLAB_m1","treemix/poplist",cex = 0.75)
t1<-plot_tree("treemix/migrations/fwsw_FLAB_m1",scale=T,mbar=T,cex = 0.75,mig_left=FALSE)

r2<-plot_resid("treemix/migrations/fwsw_FLAB_m2","treemix/poplist",cex = 0.75)
t2<-plot_tree("treemix/migrations/fwsw_FLAB_m2",scale=T,mbar=F,cex = 0.75)

r3<-plot_resid("treemix/migrations/fwsw_FLAB_m3","treemix/poplist",cex = 0.75)
t3<-plot_tree("treemix/migrations/fwsw_FLAB_m3",scale=T,mbar=F,cex = 0.75)

r4<-plot_resid("treemix/migrations/fwsw_FLAB_m4","treemix/poplist",cex = 0.75)
t4<-plot_tree("treemix/migrations/fwsw_FLAB_m4",scale=T,mbar=F,cex = 0.75)

r5<-plot_resid("treemix/migrations/fwsw_FLAB_m5","treemix/poplist",cex = 0.75)
t5<-plot_tree("treemix/migrations/fwsw_FLAB_m5",scale=T,mbar=F,cex = 0.75)

```


To choose the optimal number of migration edges, we used the R package optM [@fitakOptMPackageOptimize2019]. Using the linear analytical method from the package, we determined that the best number of migration edges was 1 (Figure \@ref(fig:plotOptM)). This finding was consistent with inspecting the log likelihoods of the bootstraps with various migration edges (\@ref(fig:treemixLlik)).

```{r optM,message=FALSE}
library(OptM)
tmOpt<-optM("treemix/migrations/")
```
```{r plotOptM,fig.keep='true',message=FALSE,fig.cap="Scree plot showing the comparison of Treemix number of migration edges."}
evanno_treemix(tmOpt)
```

```{r treemixMigrationEdges,message=FALSE}
m1s<-list.files(pattern="m1.*treeout",path="treemix/migrations/",full.names = TRUE)
edges<-do.call(rbind,lapply(m1s,function(file){
  treeout<-scan(file,what="character",sep='\n')
  edge<-do.call(rbind,strsplit(treeout[2],' '))
  return(edge)
}))
rownames(edges)<-m1s
```

```{r M1Starts}
starts<-gsub("[[:digit:]]","\\1",edges[,5])
starts<-gsub(":","",starts)
starts<-gsub("\\.","",starts)
starts<-gsub("e-","",starts)
starts<-gsub(",",", ",starts)
starTab<-data.frame(tree=names(summary(as.factor(starts))),
                    counts=summary(as.factor(starts)))
rownames(starTab)<-NULL
kable(starTab,"latex",booktabs=TRUE,row.names=FALSE,longtable=TRUE,
      col.names = c("tree location","number of bootstraps"),
      caption="Number of bootstraps with the one migration edge beginning at each of these points on the population trees.") %>%
  kable_styling(latex_options=c("HOLD_position","repeat_header"))  %>%
  column_spec(1,width = "30em")
```

```{r M1Stops}
stops<-gsub("[[:digit:]]","\\1",edges[,6])
stops<-gsub(":","",stops)
stops<-gsub("\\.","",stops)
stops<-gsub("e-","",stops)
kable(summary(as.factor(stops)),"latex",booktabs=TRUE,
      col.names = c("number of bootstraps"),
      caption="Number of bootstraps with the one migration edge ending at each of these points on the population trees.")%>%
  kable_styling(latex_options=c("HOLD_position"))
```

The overwhelming majority of bootstrapped trees with one migration edge have that edge leading from the main branch of the tree to FLPB (see Tables \@ref(tab:M1Starts) and \@ref(tab:M1Stops) for the common start and end points). 

The model chosen with the optimal number of migration edges had a similar tree structure to the rooted tree with no migration edges, with the Atlantic coast of Florida, the Gulf coast of Florida forming another cluster, the Alabama and Louisiana populations, and the Texas populations each forming their own clusters, and FLLG (which is the Florida freshwater population) branching off between the two Florida clusters (Figure \@ref(fig:treemixCompare)). The two migration edges run from the branch leading to the Gulf coast of Florida populations to the FLAB population and from the FLSI-FLAB node to the node between the TXFW population and the other Texas populations (Figure \@ref(fig:treemixCompare)). 

```{r MaketreemixCompare,results='hide',eval=TRUE}
bestM1<-gsub("\\.llik","",
             rownames(likes[likes$migs==1,])[
               which.max(likes$loglikelihood[likes$migs==1])])
png("../figs/treemix_comparison.png",height = 4,width=8,units="in",res=300)
par(mfrow=c(1,3),mar=c(1,2,1,2),oma=c(1,6,1,2),xpd=TRUE)
t0<-plot_tree("treemix/migrations/fwsw_FLAB_m0",scale=T,mbar=F,cex = 1.5,
              lwd=2,disp=0.0002,scadj=0.05)
t1<-plot_tree("treemix/migrations/fwsw_FLAB_m1",scale=T,mbar=T,cex = 1.5,
              lwd=2,mig_left=FALSE,disp=0.0002,scadj=0.05)
t2<-plot_tree(bestM1,scale=T,mbar=T,cex = 1.5,
              lwd=2,mig_left=FALSE,disp=0.0002,scadj=0.05)
dev.off()
```

```{r treemixCompare, out.width='90%',fig.cap="The plot of the tree with FLAB as root but no migration edges (left) compared to the tree with FLAB as root and two migration edges. The drift parameter is plotted on the x-axis, and migration edges are colored based on the migration weight."}
knitr::include_graphics('../figs/treemix_comparison.png')
```



### Threepop and fourpop analysis
 
We investigated the two migration edges that are in the best Treemix model using threepop and fourpop analyses. In the threepop analysis, significantly negative f3 statistics mean that the first pop in the list (A in A;B,C) is admixed. Therefore, with the threepop analysis we want to look for the tree (A;B,C) where A corresponds to the end of an arrow and (B,C) corresponds to where an arrow begins. In the fourpop analysis, a significantly non-zero value indicates gene flow in the tree.

```{r readf3f4}
threepop<-data.frame(do.call(rbind,
                             strsplit(grep(
                               ";",readLines("treemix/fwsw_threepop.txt"),
                               value = TRUE),' ')))
colnames(threepop)<-c("pops","f3_stat","f3_se","f3_z")
threepop$f3_p<-pnorm(-abs(as.numeric(as.character(threepop$f3_z)))) # one sided

fourpop<-data.frame(do.call(rbind,
                            strsplit(grep(";",
                                          readLines("treemix/fwsw_fourpop.txt"),
                                          value = TRUE),' ')))
colnames(fourpop)<-c("pops","f4_stat","f4_se","f4_z")
fourpop$f4_p<-2*pnorm(-abs(as.numeric(as.character(fourpop$f4_z))))
```

#### Migration Edge 1

The first migration edge we investigated indicated potential migration from the FLSI-FLAB node to the TXFW-other Texas populations node. We first investigated all of the three-population trees with TXFW and FLSI, and found that none of these f3 statistics are zero (Table \@ref(tab:f3Edge1)), which suggests that TXFW is not admixed with FLSI. 

```{r f3Edge1,fig.pos='H'}
kable(threepop[grep("TXFW;.*FLSI",threepop$pops),],"latex",booktabs=TRUE,
      caption = "Three-population trees with TXFW and FLSI and their f3 statistic, standard error, z-score, and p-value.") %>%
  kable_styling(latex_options="HOLD_position")
```

To narrow our focus, we extracted the threepop analyses for the trees (TXFW;TXCB,FLSI) and (TXFW,FLSI,FLAB), and again found non-zero f3 statistics (Table \@ref(tab:f3Edge1Specific)).

```{r f3Edge1Specific,fig.pos='H'}
kable(threepop[threepop$pops=="TXFW;TXCB,FLSI" | 
           threepop$pops=="TXFW;FLSI,FLAB",] ,"latex",booktabs=TRUE,
      caption = "The two most relevant three-population trees with TXFW and FLSI and their f3 statistic, standard error, z-score, and p-value.")%>%
  kable_styling(latex_options="HOLD_position")
```

To investigate the results of the fourpop analysis for this migration edge, we focused on four-population trees including the TXFW,TXCB population pair and FLSI (Table \@ref(tab:f4Edge1)). The f4 statistics were very close to zero and most have standard errors overlapping zero, indicating that gene flow is not particularly likely.

```{r f4Edge1,fig.pos='H'}
kable(fourpop[grep("TXFW,TXCB;.*FLSI.*",fourpop$pops),],"latex",booktabs=TRUE,
      caption = "Four-population trees with TXFW, TXCB, and FLSI and their f4 statistic, standard error, z-score, and p-value." )%>%
  kable_styling(latex_options="HOLD_position")
```

Focusing on the most relevant fourpop analysis, with the tree (TXFW,TXCB;FLSI,FLAB), we can see that the f4 statistic is slightly negative and overlaps with zero (Table \@ref(tab:f4Edge1Specific)), suggesting that current gene flow is not likely between these branches.

```{r f4Edge1Specific,fig.pos='H'}
kable(fourpop[fourpop$pops=="TXFW,TXCB;FLSI,FLAB",] ,"latex",booktabs=TRUE,
      caption = "The most relevant four-population tree with TXFW, TXCB, and FLSI and its f4 statistic, standard error, z-score, and p-value.")%>%
  kable_styling(latex_options="HOLD_position")
```

#### Migration Edge 2

The second migration edge was indicated potential migratoin from the branch leading to the Gulf of Mexico populations to the FLAB population. We therefore began by investigating all three-population trees involving FLAB, and then narrowed our focus to the (FLAB;FLPB,FLFW) tree. 

```{r f3Edge2,fig.pos='H'}
kable(threepop[grep("FLAB;.*",threepop$pops),],"latex",booktabs=TRUE,longtable=TRUE,row.names=FALSE,
      caption = "Three-population trees with FLAB their f3 statistic, standard error, z-score, and p-value.") %>%
  kable_styling(latex_options="HOLD_position")
```

```{r f3Edge2Specific,fig.pos='H'}
kable(threepop[threepop$pops=="FLAB;FLPB,FLLG",],"latex",booktabs=TRUE,
      caption = "Three-population tree with FLAB, FLPB, and FLFW and its f3 statistic, standard error, z-score, and p-value.")%>%
  kable_styling(latex_options="HOLD_position","repeat_header")
```

Inspecting this three-population tree (Table \@ref(tab:f3Edge2Specific)) shows that the statistic is positive does not overlap with zero, and none of the FLAB values are negative, suggesting that this migration edge is unlikely to represent admixture. 


We then investigated the fourpop analysis for this migration edge, focusing on the FLAB population and any other populations in Florida. All of these f4 statistics are very close to zero  (Table \@ref(tab:f4Edge2)), which suggests that there is not much evidence for gene flow between these branches. We narrowed our focus to the FLAB,FLCC;FLPB,FLFW tree. This tree's f4 statistic is negative but overlaps zero  (Table \@ref(tab:f4Edge2best)), so is also not indicative of gene flow.

```{r f4Edge2,fig.pos='H'}
kable(fourpop[grep("FLAB,FL.*;.*",fourpop$pops),],format="latex",booktabs=TRUE,
      caption = "Four-population trees with FLAB and one other Florida population and their f4 statistic, standard error, z-score, and p-value." )%>%
  kable_styling(latex_options=c("HOLD_position"))

```

```{r f4Edge2best,fig.pos='H'}
kable(fourpop[fourpop$pops=="FLAB,FLCC;FLPB,FLLG",],"latex",booktabs=TRUE,
      caption = "The most relevant four-population tree with TXFW, TXCB, and FLSI and its f4 statistic, standard error, z-score, and p-value.")%>%
  kable_styling(latex_options="HOLD_position")
```


Therefore, we concluded that these migration edges are indicative of historical processes connecting these geographically distinct populations, rather than direct, contemporary gene flow. 




# References {-}


